{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import gseapy as gp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgl_genes = pd.read_csv(\"/mnt/c/Users/ouyangkang/Downloads/microglia_markers.csv\")\n",
    "mac_genes = pd.read_csv(\"/mnt/c/Users/ouyangkang/Downloads/macrophage_markers.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29, 10)\n",
      "(158, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>idx</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>FDR</th>\n",
       "      <th>MeanDiff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18092</td>\n",
       "      <td>chr6</td>\n",
       "      <td>108489743</td>\n",
       "      <td>108489803</td>\n",
       "      <td>1</td>\n",
       "      <td>Mir7661</td>\n",
       "      <td>805</td>\n",
       "      <td>1.790966</td>\n",
       "      <td>9.062958e-46</td>\n",
       "      <td>1.202686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18093</td>\n",
       "      <td>chr6</td>\n",
       "      <td>108660934</td>\n",
       "      <td>108577036</td>\n",
       "      <td>2</td>\n",
       "      <td>0610040F04Rik</td>\n",
       "      <td>806</td>\n",
       "      <td>1.312830</td>\n",
       "      <td>1.023321e-39</td>\n",
       "      <td>0.710494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18240</td>\n",
       "      <td>chr6</td>\n",
       "      <td>122626424</td>\n",
       "      <td>122630271</td>\n",
       "      <td>1</td>\n",
       "      <td>Dppa3</td>\n",
       "      <td>953</td>\n",
       "      <td>1.660713</td>\n",
       "      <td>3.140691e-39</td>\n",
       "      <td>0.790734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18241</td>\n",
       "      <td>chr6</td>\n",
       "      <td>122707489</td>\n",
       "      <td>122714633</td>\n",
       "      <td>1</td>\n",
       "      <td>Nanog</td>\n",
       "      <td>954</td>\n",
       "      <td>1.576998</td>\n",
       "      <td>4.150824e-38</td>\n",
       "      <td>0.455416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>743</td>\n",
       "      <td>chr1</td>\n",
       "      <td>128592299</td>\n",
       "      <td>128588199</td>\n",
       "      <td>2</td>\n",
       "      <td>Cxcr4</td>\n",
       "      <td>743</td>\n",
       "      <td>1.639783</td>\n",
       "      <td>2.213639e-27</td>\n",
       "      <td>1.046892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 seqnames      start        end  strand           name  idx  \\\n",
       "0       18092     chr6  108489743  108489803       1        Mir7661  805   \n",
       "1       18093     chr6  108660934  108577036       2  0610040F04Rik  806   \n",
       "2       18240     chr6  122626424  122630271       1          Dppa3  953   \n",
       "3       18241     chr6  122707489  122714633       1          Nanog  954   \n",
       "4         743     chr1  128592299  128588199       2          Cxcr4  743   \n",
       "\n",
       "     Log2FC           FDR  MeanDiff  \n",
       "0  1.790966  9.062958e-46  1.202686  \n",
       "1  1.312830  1.023321e-39  0.710494  \n",
       "2  1.660713  3.140691e-39  0.790734  \n",
       "3  1.576998  4.150824e-38  0.455416  \n",
       "4  1.639783  2.213639e-27  1.046892  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(mgl_genes.shape)\n",
    "print(mac_genes.shape)\n",
    "mac_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_10x_scRNA_2021',\n",
       " 'Allen_Brain_Atlas_down',\n",
       " 'Allen_Brain_Atlas_up',\n",
       " 'Azimuth_2023',\n",
       " 'Azimuth_Cell_Types_2021',\n",
       " 'BioCarta_2013',\n",
       " 'BioCarta_2015',\n",
       " 'BioCarta_2016',\n",
       " 'BioPlanet_2019',\n",
       " 'BioPlex_2017',\n",
       " 'CCLE_Proteomics_2020',\n",
       " 'CORUM',\n",
       " 'COVID-19_Related_Gene_Sets',\n",
       " 'COVID-19_Related_Gene_Sets_2021',\n",
       " 'Cancer_Cell_Line_Encyclopedia',\n",
       " 'CellMarker_2024',\n",
       " 'CellMarker_Augmented_2021',\n",
       " 'ChEA_2013',\n",
       " 'ChEA_2015',\n",
       " 'ChEA_2016',\n",
       " 'ChEA_2022',\n",
       " 'Chromosome_Location',\n",
       " 'Chromosome_Location_hg19',\n",
       " 'ClinVar_2019',\n",
       " 'DGIdb_Drug_Targets_2024',\n",
       " 'DSigDB',\n",
       " 'Data_Acquisition_Method_Most_Popular_Genes',\n",
       " 'DepMap_CRISPR_GeneDependency_CellLines_2023',\n",
       " 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
       " 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
       " 'Descartes_Cell_Types_and_Tissue_2021',\n",
       " 'Diabetes_Perturbations_GEO_2022',\n",
       " 'DisGeNET',\n",
       " 'Disease_Perturbations_from_GEO_down',\n",
       " 'Disease_Perturbations_from_GEO_up',\n",
       " 'Disease_Signatures_from_GEO_down_2014',\n",
       " 'Disease_Signatures_from_GEO_up_2014',\n",
       " 'DrugMatrix',\n",
       " 'Drug_Perturbations_from_GEO_2014',\n",
       " 'Drug_Perturbations_from_GEO_down',\n",
       " 'Drug_Perturbations_from_GEO_up',\n",
       " 'ENCODE_Histone_Modifications_2013',\n",
       " 'ENCODE_Histone_Modifications_2015',\n",
       " 'ENCODE_TF_ChIP-seq_2014',\n",
       " 'ENCODE_TF_ChIP-seq_2015',\n",
       " 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
       " 'ESCAPE',\n",
       " 'Elsevier_Pathway_Collection',\n",
       " 'Enrichr_Libraries_Most_Popular_Genes',\n",
       " 'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
       " 'Enrichr_Users_Contributed_Lists_2020',\n",
       " 'Epigenomics_Roadmap_HM_ChIP-seq',\n",
       " 'FANTOM6_lncRNA_KD_DEGs',\n",
       " 'GO_Biological_Process_2013',\n",
       " 'GO_Biological_Process_2015',\n",
       " 'GO_Biological_Process_2017',\n",
       " 'GO_Biological_Process_2017b',\n",
       " 'GO_Biological_Process_2018',\n",
       " 'GO_Biological_Process_2021',\n",
       " 'GO_Biological_Process_2023',\n",
       " 'GO_Cellular_Component_2013',\n",
       " 'GO_Cellular_Component_2015',\n",
       " 'GO_Cellular_Component_2017',\n",
       " 'GO_Cellular_Component_2017b',\n",
       " 'GO_Cellular_Component_2018',\n",
       " 'GO_Cellular_Component_2021',\n",
       " 'GO_Cellular_Component_2023',\n",
       " 'GO_Molecular_Function_2013',\n",
       " 'GO_Molecular_Function_2015',\n",
       " 'GO_Molecular_Function_2017',\n",
       " 'GO_Molecular_Function_2017b',\n",
       " 'GO_Molecular_Function_2018',\n",
       " 'GO_Molecular_Function_2021',\n",
       " 'GO_Molecular_Function_2023',\n",
       " 'GTEx_Aging_Signatures_2021',\n",
       " 'GTEx_Tissue_Expression_Down',\n",
       " 'GTEx_Tissue_Expression_Up',\n",
       " 'GTEx_Tissues_V8_2023',\n",
       " 'GWAS_Catalog_2019',\n",
       " 'GWAS_Catalog_2023',\n",
       " 'GeDiPNet_2023',\n",
       " 'GeneSigDB',\n",
       " 'Gene_Perturbations_from_GEO_down',\n",
       " 'Gene_Perturbations_from_GEO_up',\n",
       " 'Genes_Associated_with_NIH_Grants',\n",
       " 'Genome_Browser_PWMs',\n",
       " 'GlyGen_Glycosylated_Proteins_2022',\n",
       " 'HDSigDB_Human_2021',\n",
       " 'HDSigDB_Mouse_2021',\n",
       " 'HMDB_Metabolites',\n",
       " 'HMS_LINCS_KinomeScan',\n",
       " 'HomoloGene',\n",
       " 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
       " 'HuBMAP_ASCTplusB_augmented_2022',\n",
       " 'HumanCyc_2015',\n",
       " 'HumanCyc_2016',\n",
       " 'Human_Gene_Atlas',\n",
       " 'Human_Phenotype_Ontology',\n",
       " 'IDG_Drug_Targets_2022',\n",
       " 'InterPro_Domains_2019',\n",
       " 'Jensen_COMPARTMENTS',\n",
       " 'Jensen_DISEASES',\n",
       " 'Jensen_TISSUES',\n",
       " 'KEA_2013',\n",
       " 'KEA_2015',\n",
       " 'KEGG_2013',\n",
       " 'KEGG_2015',\n",
       " 'KEGG_2016',\n",
       " 'KEGG_2019_Human',\n",
       " 'KEGG_2019_Mouse',\n",
       " 'KEGG_2021_Human',\n",
       " 'KOMP2_Mouse_Phenotypes_2022',\n",
       " 'Kinase_Perturbations_from_GEO_down',\n",
       " 'Kinase_Perturbations_from_GEO_up',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_down',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_up',\n",
       " 'LINCS_L1000_CRISPR_KO_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_down',\n",
       " 'LINCS_L1000_Chem_Pert_up',\n",
       " 'LINCS_L1000_Ligand_Perturbations_down',\n",
       " 'LINCS_L1000_Ligand_Perturbations_up',\n",
       " 'Ligand_Perturbations_from_GEO_down',\n",
       " 'Ligand_Perturbations_from_GEO_up',\n",
       " 'MAGMA_Drugs_and_Diseases',\n",
       " 'MAGNET_2023',\n",
       " 'MCF7_Perturbations_from_GEO_down',\n",
       " 'MCF7_Perturbations_from_GEO_up',\n",
       " 'MGI_Mammalian_Phenotype_2013',\n",
       " 'MGI_Mammalian_Phenotype_2017',\n",
       " 'MGI_Mammalian_Phenotype_Level_3',\n",
       " 'MGI_Mammalian_Phenotype_Level_4',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2019',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2021',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2024',\n",
       " 'MSigDB_Computational',\n",
       " 'MSigDB_Hallmark_2020',\n",
       " 'MSigDB_Oncogenic_Signatures',\n",
       " 'Metabolomics_Workbench_Metabolites_2022',\n",
       " 'Microbe_Perturbations_from_GEO_down',\n",
       " 'Microbe_Perturbations_from_GEO_up',\n",
       " 'MoTrPAC_2023',\n",
       " 'Mouse_Gene_Atlas',\n",
       " 'NCI-60_Cancer_Cell_Lines',\n",
       " 'NCI-Nature_2016',\n",
       " 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
       " 'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
       " 'NURSA_Human_Endogenous_Complexome',\n",
       " 'OMIM_Disease',\n",
       " 'OMIM_Expanded',\n",
       " 'Old_CMAP_down',\n",
       " 'Old_CMAP_up',\n",
       " 'Orphanet_Augmented_2021',\n",
       " 'PFOCR_Pathways',\n",
       " 'PFOCR_Pathways_2023',\n",
       " 'PPI_Hub_Proteins',\n",
       " 'PanglaoDB_Augmented_2021',\n",
       " 'Panther_2015',\n",
       " 'Panther_2016',\n",
       " 'Pfam_Domains_2019',\n",
       " 'Pfam_InterPro_Domains',\n",
       " 'PheWeb_2019',\n",
       " 'PhenGenI_Association_2021',\n",
       " 'Phosphatase_Substrates_from_DEPOD',\n",
       " 'ProteomicsDB_2020',\n",
       " 'Proteomics_Drug_Atlas_2023',\n",
       " 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
       " 'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_AutoRIF_Gene_Lists',\n",
       " 'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_GeneRIF_Gene_Lists',\n",
       " 'Reactome_2013',\n",
       " 'Reactome_2015',\n",
       " 'Reactome_2016',\n",
       " 'Reactome_2022',\n",
       " 'Rummagene_kinases',\n",
       " 'Rummagene_signatures',\n",
       " 'Rummagene_transcription_factors',\n",
       " 'SILAC_Phosphoproteomics',\n",
       " 'SubCell_BarCode',\n",
       " 'SynGO_2022',\n",
       " 'SynGO_2024',\n",
       " 'SysMyo_Muscle_Gene_Sets',\n",
       " 'TF-LOF_Expression_from_GEO',\n",
       " 'TF_Perturbations_Followed_by_Expression',\n",
       " 'TG_GATES_2020',\n",
       " 'TRANSFAC_and_JASPAR_PWMs',\n",
       " 'TRRUST_Transcription_Factors_2019',\n",
       " 'Table_Mining_of_CRISPR_Studies',\n",
       " 'Tabula_Muris',\n",
       " 'Tabula_Sapiens',\n",
       " 'TargetScan_microRNA',\n",
       " 'TargetScan_microRNA_2017',\n",
       " 'The_Kinase_Library_2023',\n",
       " 'The_Kinase_Library_2024',\n",
       " 'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
       " 'Tissue_Protein_Expression_from_ProteomicsDB',\n",
       " 'Transcription_Factor_PPIs',\n",
       " 'UK_Biobank_GWAS_v1',\n",
       " 'Virus-Host_PPI_P-HIPSTer_2020',\n",
       " 'VirusMINT',\n",
       " 'Virus_Perturbations_from_GEO_down',\n",
       " 'Virus_Perturbations_from_GEO_up',\n",
       " 'WikiPathway_2021_Human',\n",
       " 'WikiPathway_2023_Human',\n",
       " 'WikiPathways_2013',\n",
       " 'WikiPathways_2015',\n",
       " 'WikiPathways_2016',\n",
       " 'WikiPathways_2019_Human',\n",
       " 'WikiPathways_2019_Mouse',\n",
       " 'WikiPathways_2024_Human',\n",
       " 'WikiPathways_2024_Mouse',\n",
       " 'dbGaP',\n",
       " 'huMAP',\n",
       " 'lncHUB_lncRNA_Co-Expression',\n",
       " 'miRTarBase_2017']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp.get_library_name(organism='Human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrichment(gene_list, file_name_prefix):\n",
    "    # enrichment analysis\n",
    "    enr = gp.enrich(gene_list,\n",
    "                    gene_sets=\"GO_Biological_Process_2023\",\n",
    "                    outdir=None)\n",
    "    # filtering p value is more than 0.05\n",
    "    result = enr.res2d[enr.res2d['P-value'] < 0.05]\n",
    "    # save result\n",
    "    result.to_csv(f\"/mnt/c/Users/ouyangkang/Desktop/{file_name_prefix}_result.csv\", index=None)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgl_result = enrichment(mgl_genes.name.tolist(), 'mgl_diff')\n",
    "mac_result = enrichment(mac_genes.name.tolist(), 'mac_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Amylin Receptor Signaling Pathway (GO:0097647)</td>\n",
       "      <td>1/5</td>\n",
       "      <td>0.007230</td>\n",
       "      <td>0.12275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>178.276786</td>\n",
       "      <td>878.827148</td>\n",
       "      <td>RAMP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Vascular Associated Smooth Muscle Cell Develop...</td>\n",
       "      <td>1/6</td>\n",
       "      <td>0.008670</td>\n",
       "      <td>0.12275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>142.614286</td>\n",
       "      <td>677.124522</td>\n",
       "      <td>RAMP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Calcitonin Family Receptor Signaling Pathway (...</td>\n",
       "      <td>1/6</td>\n",
       "      <td>0.008670</td>\n",
       "      <td>0.12275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>142.614286</td>\n",
       "      <td>677.124522</td>\n",
       "      <td>RAMP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Astral Microtubule Organization (GO:0030953)</td>\n",
       "      <td>1/8</td>\n",
       "      <td>0.011543</td>\n",
       "      <td>0.12275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101.857143</td>\n",
       "      <td>454.451827</td>\n",
       "      <td>TPPP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Vascular Associated Smooth Muscle Cell Differe...</td>\n",
       "      <td>1/8</td>\n",
       "      <td>0.011543</td>\n",
       "      <td>0.12275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101.857143</td>\n",
       "      <td>454.451827</td>\n",
       "      <td>RAMP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_set  \\\n",
       "0  GO_Biological_Process_2023   \n",
       "1  GO_Biological_Process_2023   \n",
       "2  GO_Biological_Process_2023   \n",
       "3  GO_Biological_Process_2023   \n",
       "4  GO_Biological_Process_2023   \n",
       "\n",
       "                                                Term Overlap   P-value  \\\n",
       "0     Amylin Receptor Signaling Pathway (GO:0097647)     1/5  0.007230   \n",
       "1  Vascular Associated Smooth Muscle Cell Develop...     1/6  0.008670   \n",
       "2  Calcitonin Family Receptor Signaling Pathway (...     1/6  0.008670   \n",
       "3       Astral Microtubule Organization (GO:0030953)     1/8  0.011543   \n",
       "4  Vascular Associated Smooth Muscle Cell Differe...     1/8  0.011543   \n",
       "\n",
       "   Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "0           0.12275            0                     0  178.276786   \n",
       "1           0.12275            0                     0  142.614286   \n",
       "2           0.12275            0                     0  142.614286   \n",
       "3           0.12275            0                     0  101.857143   \n",
       "4           0.12275            0                     0  101.857143   \n",
       "\n",
       "   Combined Score  Genes  \n",
       "0      878.827148  RAMP2  \n",
       "1      677.124522  RAMP2  \n",
       "2      677.124522  RAMP2  \n",
       "3      454.451827   TPPP  \n",
       "4      454.451827  RAMP2  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgl_result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_657/870613516.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  select_mgl[\"Lable\"] = \"MGL\"\n",
      "/tmp/ipykernel_657/870613516.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  select_mac[\"Lable\"] = \"MAC\"\n"
     ]
    }
   ],
   "source": [
    "select_mgl = mgl_result.iloc[[0, 1, 2 ,5, 8, 14, 15, 18, 19, 22]]\n",
    "select_mgl[\"Lable\"] = \"MGL\"\n",
    "select_mac = mac_result.iloc[[0, 1, 2, 3, 5, 6, 8, 16, 24, 39]]\n",
    "select_mac[\"Lable\"] = \"MAC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Lable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Inflammatory Response</td>\n",
       "      <td>9/236</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.075060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.219377</td>\n",
       "      <td>47.511134</td>\n",
       "      <td>IL22;HP;FPR1;CYBB;CXCR4;CXCL1;ITGAL;CXCL2;S100A8</td>\n",
       "      <td>MAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Neutrophil Chemotaxis</td>\n",
       "      <td>5/70</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>0.075060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9.943188</td>\n",
       "      <td>83.252182</td>\n",
       "      <td>LGALS3;SAA1;CXCL1;CXCL2;S100A8</td>\n",
       "      <td>MAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Granulocyte Chemotaxis</td>\n",
       "      <td>5/73</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.075060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9.503076</td>\n",
       "      <td>77.695580</td>\n",
       "      <td>LGALS3;SAA1;CXCL1;CXCL2;S100A8</td>\n",
       "      <td>MAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Neutrophil Migration</td>\n",
       "      <td>5/77</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>0.075060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.973312</td>\n",
       "      <td>71.131523</td>\n",
       "      <td>LGALS3;SAA1;CXCL1;CXCL2;S100A8</td>\n",
       "      <td>MAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Histidine Metabolic Process</td>\n",
       "      <td>2/6</td>\n",
       "      <td>0.000911</td>\n",
       "      <td>0.126335</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63.583333</td>\n",
       "      <td>445.140237</td>\n",
       "      <td>HAL;HDC</td>\n",
       "      <td>MAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_set                          Term Overlap   P-value  \\\n",
       "0  GO_Biological_Process_2023        Inflammatory Response    9/236  0.000111   \n",
       "1  GO_Biological_Process_2023        Neutrophil Chemotaxis     5/70  0.000231   \n",
       "2  GO_Biological_Process_2023       Granulocyte Chemotaxis     5/73  0.000281   \n",
       "3  GO_Biological_Process_2023         Neutrophil Migration     5/77  0.000361   \n",
       "4  GO_Biological_Process_2023  Histidine Metabolic Process      2/6  0.000911   \n",
       "\n",
       "   Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
       "0          0.075060            0                     0    5.219377   \n",
       "1          0.075060            0                     0    9.943188   \n",
       "2          0.075060            0                     0    9.503076   \n",
       "3          0.075060            0                     0    8.973312   \n",
       "4          0.126335            0                     0   63.583333   \n",
       "\n",
       "   Combined Score                                             Genes Lable  \n",
       "0       47.511134  IL22;HP;FPR1;CYBB;CXCR4;CXCL1;ITGAL;CXCL2;S100A8   MAC  \n",
       "1       83.252182                    LGALS3;SAA1;CXCL1;CXCL2;S100A8   MAC  \n",
       "2       77.695580                    LGALS3;SAA1;CXCL1;CXCL2;S100A8   MAC  \n",
       "3       71.131523                    LGALS3;SAA1;CXCL1;CXCL2;S100A8   MAC  \n",
       "4      445.140237                                           HAL;HDC   MAC  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_all = pd.concat([select_mac, select_mgl], ignore_index=True)\n",
    "select_all.Term = select_all.Term.str.split(\"(\").str[0]\n",
    "select_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ouyangkang/software/miniconda3/envs/sc/lib/python3.11/site-packages/gseapy/plot.py:694: FutureWarning: The 'method' keyword in Series.replace is deprecated and will be removed in a future version.\n",
      "  df[self.colname].replace(\n",
      "/home/ouyangkang/software/miniconda3/envs/sc/lib/python3.11/site-packages/gseapy/plot.py:694: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  df[self.colname].replace(\n"
     ]
    }
   ],
   "source": [
    "gp.barplot(\n",
    "    select_all,\n",
    "    figsize=(6, 8),\n",
    "    title=\"Different function of DEGs between MGL and MAC\",\n",
    "    group=\"Lable\",\n",
    "    color=[\"b\", \"r\"],\n",
    "    column=\"P-value\",\n",
    "    ofname = \"MGL-MAC_diff_function.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
