{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Host</th>\n",
       "      <th>Virus</th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>VirusTaxID</th>\n",
       "      <th>HostNCBIResolved</th>\n",
       "      <th>VirusNCBIResolved</th>\n",
       "      <th>ICTVRatified</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>...</th>\n",
       "      <th>PMID</th>\n",
       "      <th>ReleaseYear</th>\n",
       "      <th>ReleaseMonth</th>\n",
       "      <th>ReleaseDay</th>\n",
       "      <th>CollectionYear</th>\n",
       "      <th>CollectionMonth</th>\n",
       "      <th>CollectionDay</th>\n",
       "      <th>AssocID</th>\n",
       "      <th>DatabaseDOI</th>\n",
       "      <th>NCBIAccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus btcov/ncl_mco1/mop_con/rsa/2012</td>\n",
       "      <td>258862</td>\n",
       "      <td>1532174</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>mops</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KF843853.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus p.sp/k/spain/2007</td>\n",
       "      <td>27671</td>\n",
       "      <td>1076425</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HQ184060.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus sp.</td>\n",
       "      <td>27671</td>\n",
       "      <td>1906673</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LC334391.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus tadaridam4</td>\n",
       "      <td>9437</td>\n",
       "      <td>1582306</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>tadarida</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KM514667.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>avian coronavirus</td>\n",
       "      <td>53274</td>\n",
       "      <td>694014</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>rhizomys</td>\n",
       "      <td>spalacidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021</td>\n",
       "      <td>6.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT222079.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35606</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>avian coronavirus</td>\n",
       "      <td>30538</td>\n",
       "      <td>694014</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005</td>\n",
       "      <td>3.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AY939836.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35607</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>avian coronavirus</td>\n",
       "      <td>30538</td>\n",
       "      <td>694014</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AY945840.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35608</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>betacoronavirus 1</td>\n",
       "      <td>30538</td>\n",
       "      <td>694003</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>218960.0</td>\n",
       "      <td>https://doi.org/10.1038/sdata.2015.49</td>\n",
       "      <td>124294673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35609</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>betacoronavirus 1</td>\n",
       "      <td>30538</td>\n",
       "      <td>694003</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2006</td>\n",
       "      <td>9.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DQ915164.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35610</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>human coronavirus 229e</td>\n",
       "      <td>30538</td>\n",
       "      <td>11137</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2012</td>\n",
       "      <td>4.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JQ410000.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35611 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Host                                             Virus  \\\n",
       "0                NaN  alphacoronavirus btcov/ncl_mco1/mop_con/rsa/2012   \n",
       "1                NaN                alphacoronavirus p.sp/k/spain/2007   \n",
       "2                NaN                              alphacoronavirus sp.   \n",
       "3                NaN                       alphacoronavirus tadaridam4   \n",
       "4                NaN                                 avian coronavirus   \n",
       "...              ...                                               ...   \n",
       "35606  vicugna pacos                                 avian coronavirus   \n",
       "35607  vicugna pacos                                 avian coronavirus   \n",
       "35608  vicugna pacos                                 betacoronavirus 1   \n",
       "35609  vicugna pacos                                 betacoronavirus 1   \n",
       "35610  vicugna pacos                            human coronavirus 229e   \n",
       "\n",
       "       HostTaxID  VirusTaxID  HostNCBIResolved  VirusNCBIResolved  \\\n",
       "0         258862     1532174              True               True   \n",
       "1          27671     1076425              True               True   \n",
       "2          27671     1906673              True               True   \n",
       "3           9437     1582306              True               True   \n",
       "4          53274      694014              True               True   \n",
       "...          ...         ...               ...                ...   \n",
       "35606      30538      694014              True               True   \n",
       "35607      30538      694014              True               True   \n",
       "35608      30538      694003              True               True   \n",
       "35609      30538      694003              True               True   \n",
       "35610      30538       11137              True               True   \n",
       "\n",
       "       ICTVRatified     HostGenus        HostFamily     HostOrder  ... PMID  \\\n",
       "0             False          mops        molossidae    chiroptera  ...  NaN   \n",
       "1             False  pipistrellus  vespertilionidae    chiroptera  ...  NaN   \n",
       "2             False  pipistrellus  vespertilionidae    chiroptera  ...  NaN   \n",
       "3             False      tadarida        molossidae    chiroptera  ...  NaN   \n",
       "4              True      rhizomys        spalacidae      rodentia  ...  NaN   \n",
       "...             ...           ...               ...           ...  ...  ...   \n",
       "35606          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35607          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35608          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35609          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35610          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "\n",
       "      ReleaseYear ReleaseMonth ReleaseDay CollectionYear CollectionMonth  \\\n",
       "0            2014         12.0       12.0         2012.0             NaN   \n",
       "1            2011          8.0       21.0         2007.0             NaN   \n",
       "2            2017         11.0       28.0            NaN             NaN   \n",
       "3            2014         12.0       17.0         2011.0             NaN   \n",
       "4            2021          6.0       23.0         2014.0             3.0   \n",
       "...           ...          ...        ...            ...             ...   \n",
       "35606        2005          3.0       21.0            NaN             NaN   \n",
       "35607        2005          3.0       22.0            NaN             NaN   \n",
       "35608        2006          NaN        NaN            NaN             NaN   \n",
       "35609        2006          9.0       30.0            NaN             NaN   \n",
       "35610        2012          4.0       29.0         2008.0             NaN   \n",
       "\n",
       "      CollectionDay   AssocID                            DatabaseDOI  \\\n",
       "0               NaN       NaN                                    NaN   \n",
       "1               NaN       NaN                                    NaN   \n",
       "2               NaN       NaN                                    NaN   \n",
       "3               NaN       NaN                                    NaN   \n",
       "4              19.0       NaN                                    NaN   \n",
       "...             ...       ...                                    ...   \n",
       "35606           NaN       NaN                                    NaN   \n",
       "35607           NaN       NaN                                    NaN   \n",
       "35608           NaN  218960.0  https://doi.org/10.1038/sdata.2015.49   \n",
       "35609           NaN       NaN                                    NaN   \n",
       "35610           NaN       NaN                                    NaN   \n",
       "\n",
       "      NCBIAccession  \n",
       "0        KF843853.1  \n",
       "1        HQ184060.1  \n",
       "2        LC334391.1  \n",
       "3        KM514667.1  \n",
       "4        MT222079.1  \n",
       "...             ...  \n",
       "35606    AY939836.2  \n",
       "35607    AY945840.1  \n",
       "35608     124294673  \n",
       "35609    DQ915164.2  \n",
       "35610    JQ410000.1  \n",
       "\n",
       "[35611 rows x 34 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "# input data\n",
    "coro_data = pd.read_csv('coro_data.csv')\n",
    "coro_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-4-890880d44c55>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  id_accession_data.dropna(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# get coroviridae data with NCBIAccession number\n",
    "id_accession_data = coro_data[['VirusTaxID','NCBIAccession']]\n",
    "# remove na\n",
    "id_accession_data.dropna(inplace=True)\n",
    "\n",
    "# construct a dictionary, the key of which is ID number and value is NCBIAccession\n",
    "dic_ID_Accession = {}\n",
    "\n",
    "for key,value in zip(id_accession_data['VirusTaxID'],id_accession_data['NCBIAccession']):\n",
    "    process_value = []\n",
    "    # split the accession number by comma\n",
    "    for i in value.split(','):\n",
    "        # remove space\n",
    "        process_value.append(i.replace(\" \",\"\"))\n",
    "    dic_ID_Accession[key] = dic_ID_Accession.get(key, []) + process_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct two lists to conserve id and accession\n",
    "IDs = []\n",
    "Accessions = []\n",
    "\n",
    "# construct a two columns data frame\n",
    "for key, value in dic_ID_Accession.items():\n",
    "    n_value_count = len(value)\n",
    "    IDs = IDs + [key]*n_value_count\n",
    "    Accessions = Accessions + value\n",
    "\n",
    "coro_id_accession_csv_data = pd.DataFrame(IDs, Accessions)\n",
    "coro_id_accession_csv_data = coro_id_accession_csv_data.reset_index() # reset the column name\n",
    "coro_id_accession_csv_data.columns = ['Accession', 'VirusID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1027431, 2)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coro_id_accession_csv_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1027217, 2)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove duplication\n",
    "coro_id_accession_csv_data = coro_id_accession_csv_data.drop_duplicates(['Accession', 'VirusID'])\n",
    "\n",
    "# remoce na\n",
    "coro_id_accession_csv_data.dropna(axis=0,inplace=True)\n",
    "\n",
    "# remove empty value\n",
    "coro_id_accession_csv_data = coro_id_accession_csv_data[coro_id_accession_csv_data['Accession'] != '']\n",
    "coro_id_accession_csv_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conserve data into a excel table\n",
    "coro_id_accession_csv_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_id_accession.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input coroviridae data with VirusID and Accession\n",
    "import pandas as pd\n",
    "coro_id_accession = pd.read_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_id_accession.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the number that represents the NCBI id of url\n",
    "import re\n",
    "number = []\n",
    "accession_data = coro_id_accession[\"Accession\"]\n",
    "\n",
    "for i in accession_data:\n",
    "    pattern = re.compile(r\"^[0-9]\\d+\")\n",
    "    number = number + re.findall(pattern, i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# seperate data\n",
    "coro_id_accession = pd.read_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_id_accession.csv')\n",
    "GI_VirusID_data = coro_id_accession[coro_id_accession.Accession.isin(number)]\n",
    "Accession_VirusID_data = coro_id_accession[~(coro_id_accession.Accession.isin(number))]\n",
    "GI_VirusID_data.to_csv('GI_VirusID.csv')\n",
    "Accession_VirusID_data.to_csv('Acce_VirusID.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 87244987 id numbers are failured to convert to accession\n",
      "The 87244987 id numbers are successed to convert to accession\n",
      "The 182702146 id numbers are failured to convert to accession\n",
      "The 182702146 id numbers are successed to convert to accession\n",
      "The 226316432 id numbers are failured to convert to accession\n",
      "The 226316432 id numbers are successed to convert to accession\n"
     ]
    }
   ],
   "source": [
    "### this step can be passed\n",
    "# get the accession number through the number id\n",
    "import requests\n",
    "import re\n",
    "\n",
    "def get_fasta_info(id):\n",
    "    # get sequencing information from id\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=59cbd6ff6b3ce5d506f0eeb69e4e09362409'.format(id)\n",
    "    fasta_info = requests.get(fasta_url,timeout=(10,45)).text\n",
    "    return fasta_info\n",
    "\"\"\"\n",
    "number_accession = {}\n",
    "failured_number_id = []\n",
    "\"\"\"\n",
    "# using WHILE can improve the readablity of code\n",
    "# the step is designed to convert GI numbers to Accession number, which consumes a lot of time\n",
    "\"\"\"\n",
    "for i in range(12):\n",
    "    for id in number[i*100:(i+1)*100]:\n",
    "        try:\n",
    "            fasta_info = get_fasta_info(id)\n",
    "            pattern = re.compile(r'>(\\w+\\d+.\\d)')\n",
    "            match = re.match(pattern, fasta_info).group()\n",
    "            number_accession[id] = match.strip('>')\n",
    "        except:\n",
    "            failured_number_id.append(id)\n",
    "            print(\"The {} id numbers are failured to convert to accession\".format(id))\n",
    "            fasta_info = get_fasta_info(failured_number_id[0])\n",
    "            pattern = re.compile(r'>(\\w+\\d+.\\d)')\n",
    "            match = re.match(pattern, fasta_info).group()\n",
    "            number_accession[id] = match.strip('>')\n",
    "            print(\"The {} id numbers are successed to convert to accession\".format(id))\n",
    "            failured_number_id.pop(0)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1195"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the length\n",
    "len(number_accession.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the step, conserving the data into a csv file, is unnecessary\n",
    "numberID_accession_data = pd.DataFrame.from_dict(number_accession, orient='index')\n",
    "numberID_accession_data = numberID_accession_data.reset_index()\n",
    "numberID_accession_data.columns = ['GI', 'Accession']\n",
    "numberID_accession_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/GI_Accession.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "numberID_accession_data = pd.read_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/GI_Accession.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conserve data\n",
    "coro_id_accession.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/final_Acce_VirusID.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1029080, 2)\n",
      "(1027892, 2)\n"
     ]
    }
   ],
   "source": [
    "renew_data = pd.read_csv('final_Acce_VirusID.csv')\n",
    "print(renew_data.shape)\n",
    "renew_data.drop_duplicates(subset=['Accession', 'VirusID'], inplace=True)\n",
    "print(renew_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "eace1a3d2d18dcb98353cfeb529f20792ec64dcc10ba0bdd358326e3f660811b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
