{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-1-55398f23cfa8>:3: DtypeWarning: Columns (23,32) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  coro_data = pd.read_csv('coro_data.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Host</th>\n",
       "      <th>Virus</th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>VirusTaxID</th>\n",
       "      <th>HostNCBIResolved</th>\n",
       "      <th>VirusNCBIResolved</th>\n",
       "      <th>ICTVRatified</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>...</th>\n",
       "      <th>PMID</th>\n",
       "      <th>ReleaseYear</th>\n",
       "      <th>ReleaseMonth</th>\n",
       "      <th>ReleaseDay</th>\n",
       "      <th>CollectionYear</th>\n",
       "      <th>CollectionMonth</th>\n",
       "      <th>CollectionDay</th>\n",
       "      <th>AssocID</th>\n",
       "      <th>DatabaseDOI</th>\n",
       "      <th>NCBIAccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10114</td>\n",
       "      <td>11118</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>rattus</td>\n",
       "      <td>muridae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33554</td>\n",
       "      <td>693996</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>371929</td>\n",
       "      <td>694002</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>pitangus</td>\n",
       "      <td>tyrannidae</td>\n",
       "      <td>passeriformes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49442</td>\n",
       "      <td>694002</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>rhinolophus</td>\n",
       "      <td>rhinolophidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9397</td>\n",
       "      <td>11118</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9397</td>\n",
       "      <td>693996</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9787</td>\n",
       "      <td>694002</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>perissodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9822</td>\n",
       "      <td>693996</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>sus</td>\n",
       "      <td>suidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9895</td>\n",
       "      <td>694002</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bovidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus 1</td>\n",
       "      <td>33554</td>\n",
       "      <td>693997</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Host               Virus  HostTaxID  VirusTaxID  HostNCBIResolved  \\\n",
       "0  NaN                 NaN      10114       11118              True   \n",
       "1  NaN                 NaN      33554      693996              True   \n",
       "2  NaN                 NaN     371929      694002              True   \n",
       "3  NaN                 NaN      49442      694002              True   \n",
       "4  NaN                 NaN       9397       11118              True   \n",
       "5  NaN                 NaN       9397      693996              True   \n",
       "6  NaN                 NaN       9787      694002              True   \n",
       "7  NaN                 NaN       9822      693996              True   \n",
       "8  NaN                 NaN       9895      694002              True   \n",
       "9  NaN  alphacoronavirus 1      33554      693997              True   \n",
       "\n",
       "   VirusNCBIResolved  ICTVRatified    HostGenus     HostFamily  \\\n",
       "0               True         False       rattus        muridae   \n",
       "1               True         False          NaN            NaN   \n",
       "2               True         False     pitangus     tyrannidae   \n",
       "3               True         False  rhinolophus  rhinolophidae   \n",
       "4               True         False          NaN            NaN   \n",
       "5               True         False          NaN            NaN   \n",
       "6               True         False          NaN            NaN   \n",
       "7               True         False          sus         suidae   \n",
       "8               True         False          NaN        bovidae   \n",
       "9               True          True          NaN            NaN   \n",
       "\n",
       "        HostOrder  ... PMID ReleaseYear ReleaseMonth ReleaseDay  \\\n",
       "0        rodentia  ...  NaN         NaN          NaN        NaN   \n",
       "1       carnivora  ...  NaN         NaN          NaN        NaN   \n",
       "2   passeriformes  ...  NaN         NaN          NaN        NaN   \n",
       "3      chiroptera  ...  NaN         NaN          NaN        NaN   \n",
       "4      chiroptera  ...  NaN         NaN          NaN        NaN   \n",
       "5      chiroptera  ...  NaN         NaN          NaN        NaN   \n",
       "6  perissodactyla  ...  NaN         NaN          NaN        NaN   \n",
       "7    artiodactyla  ...  NaN         NaN          NaN        NaN   \n",
       "8    artiodactyla  ...  NaN         NaN          NaN        NaN   \n",
       "9       carnivora  ...  NaN         NaN          NaN        NaN   \n",
       "\n",
       "  CollectionYear CollectionMonth CollectionDay  AssocID DatabaseDOI  \\\n",
       "0            NaN             NaN           NaN      NaN         NaN   \n",
       "1            NaN             NaN           NaN      NaN         NaN   \n",
       "2            NaN             NaN           NaN      NaN         NaN   \n",
       "3            NaN             NaN           NaN      NaN         NaN   \n",
       "4            NaN             NaN           NaN      NaN         NaN   \n",
       "5            NaN             NaN           NaN      NaN         NaN   \n",
       "6            NaN             NaN           NaN      NaN         NaN   \n",
       "7            NaN             NaN           NaN      NaN         NaN   \n",
       "8            NaN             NaN           NaN      NaN         NaN   \n",
       "9            NaN             NaN           NaN      NaN         NaN   \n",
       "\n",
       "  NCBIAccession  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "5           NaN  \n",
       "6           NaN  \n",
       "7           NaN  \n",
       "8           NaN  \n",
       "9           NaN  \n",
       "\n",
       "[10 rows x 34 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "# input data\n",
    "coro_data = pd.read_csv('coro_data.csv')\n",
    "coro_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-2-890880d44c55>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  id_accession_data.dropna(inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# get coroviridae data with NCBIAccession number\n",
    "id_accession_data = coro_data[['VirusTaxID','NCBIAccession']]\n",
    "# remove na\n",
    "id_accession_data.dropna(inplace=True)\n",
    "\n",
    "# construct a dictionary, the key of which is ID number and value is NCBIAccession\n",
    "dic_ID_Accession = {}\n",
    "\n",
    "for key,value in zip(id_accession_data['VirusTaxID'],id_accession_data['NCBIAccession']):\n",
    "    process_value = []\n",
    "    # split the accession number by comma\n",
    "    for i in value.split(','):\n",
    "        # remove space\n",
    "        process_value.append(i.replace(\" \",\"\"))\n",
    "    dic_ID_Accession[key] = dic_ID_Accession.get(key, []) + process_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct two lists to conserve id and accession\n",
    "IDs = []\n",
    "Accessions = []\n",
    "\n",
    "# construct a two columns data frame\n",
    "for key, value in dic_ID_Accession.items():\n",
    "    n_value_count = len(value)\n",
    "    IDs = IDs + [key]*n_value_count\n",
    "    Accessions = Accessions + value\n",
    "\n",
    "coro_id_accession_csv_data = pd.DataFrame(IDs, Accessions)\n",
    "coro_id_accession_csv_data = coro_id_accession_csv_data.reset_index() # reset the column name\n",
    "coro_id_accession_csv_data.columns = ['Accession', 'VirusID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplication\n",
    "coro_id_accession_csv_data = coro_id_accession_csv_data.drop_duplicates(['Accession', 'VirusID'])\n",
    "# conserve data into a excel table\n",
    "coro_id_accession_csv_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_id_accession.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input coroviridae data with VirusID and Accession\n",
    "import pandas as pd\n",
    "coro_id_accession = pd.read_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_id_accession.csv')\n",
    "\n",
    "# forcedly convert data type from many types to string \n",
    "coro_id_accession[\"Accession\"] = coro_id_accession[\"Accession\"].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the number that represents the NCBI id of url\n",
    "import re\n",
    "number = []\n",
    "accession_data = coro_id_accession[\"Accession\"]\n",
    "\n",
    "for i in accession_data:\n",
    "    pattern = re.compile(r\"^[0-9]\\d+\")\n",
    "    number = number + re.findall(pattern, i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 87244987 id numbers are failured to convert to accession\n",
      "The 87244987 id numbers are successed to convert to accession\n",
      "The 182702146 id numbers are failured to convert to accession\n",
      "The 182702146 id numbers are successed to convert to accession\n",
      "The 226316432 id numbers are failured to convert to accession\n",
      "The 226316432 id numbers are successed to convert to accession\n"
     ]
    }
   ],
   "source": [
    "# get the accession number through the number id\n",
    "import requests\n",
    "import re\n",
    "\n",
    "def get_fasta_info(id):\n",
    "    # get sequencing information from id\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=59cbd6ff6b3ce5d506f0eeb69e4e09362409'.format(id)\n",
    "    fasta_info = requests.get(fasta_url,timeout=(10,45)).text\n",
    "    return fasta_info\n",
    "\n",
    "number_accession = {}\n",
    "failured_number_id = []\n",
    "\n",
    "# using WHILE can improve the readablity of code\n",
    "# the step is designed to convert GI numbers to Accession number, which consumes a lot of time\n",
    "for i in range(12):\n",
    "    for id in number[i*100:(i+1)*100]:\n",
    "        try:\n",
    "            fasta_info = get_fasta_info(id)\n",
    "            pattern = re.compile(r'>(\\w+\\d+.\\d)')\n",
    "            match = re.match(pattern, fasta_info).group()\n",
    "            number_accession[id] = match.strip('>')\n",
    "        except:\n",
    "            failured_number_id.append(id)\n",
    "            print(\"The {} id numbers are failured to convert to accession\".format(id))\n",
    "            fasta_info = get_fasta_info(failured_number_id[0])\n",
    "            pattern = re.compile(r'>(\\w+\\d+.\\d)')\n",
    "            match = re.match(pattern, fasta_info).group()\n",
    "            number_accession[id] = match.strip('>')\n",
    "            print(\"The {} id numbers are successed to convert to accession\".format(id))\n",
    "            failured_number_id.pop(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1195"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the length\n",
    "len(number_accession.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the step, conserving the data into a csv file, is unnecessary\n",
    "numberID_accession_data = pd.DataFrame.from_dict(number_accession, orient='index')\n",
    "numberID_accession_data = numberID_accession_data.reset_index()\n",
    "numberID_accession_data.columns = ['GI', 'Accession']\n",
    "numberID_accession_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/GI_Accession.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use dictary data format to accomplish the step is to replace GI by Accession\n",
    "coro_id_accession = coro_id_accession.replace(number_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conserve data\n",
    "coro_id_accession.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/final_Acce_VirusID.csv',index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "eace1a3d2d18dcb98353cfeb529f20792ec64dcc10ba0bdd358326e3f660811b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
