{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "data process: removing NCBI record is `FALSE` in `host` and `virus` column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>26</th>\n",
       "      <th>27</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>10114</td>\n",
       "      <td>11118</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>rattus</td>\n",
       "      <td>muridae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>33554</td>\n",
       "      <td>693996</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>371929</td>\n",
       "      <td>694002</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>pitangus</td>\n",
       "      <td>tyrannidae</td>\n",
       "      <td>passeriformes</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>49442</td>\n",
       "      <td>694002</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>rhinolophus</td>\n",
       "      <td>rhinolophidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>9397</td>\n",
       "      <td>11118</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40098</th>\n",
       "      <td>vulpes vulpes</td>\n",
       "      <td>alphacoronavirus 1</td>\n",
       "      <td>9627</td>\n",
       "      <td>693997</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>vulpes</td>\n",
       "      <td>canidae</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40099</th>\n",
       "      <td>vulpes vulpes</td>\n",
       "      <td>severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>9627</td>\n",
       "      <td>694009</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>vulpes</td>\n",
       "      <td>canidae</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td>15921605</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>219444</td>\n",
       "      <td>https://doi.org/10.1111/mec.15463</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40100</th>\n",
       "      <td>vulpes vulpes</td>\n",
       "      <td>severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>9627</td>\n",
       "      <td>694009</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>vulpes</td>\n",
       "      <td>canidae</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40101</th>\n",
       "      <td>vulpes vulpes</td>\n",
       "      <td>severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>9627</td>\n",
       "      <td>694009</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>vulpes</td>\n",
       "      <td>canidae</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40102</th>\n",
       "      <td>zalophus californianus</td>\n",
       "      <td>severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>9704</td>\n",
       "      <td>694009</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zalophus</td>\n",
       "      <td>otariidae</td>\n",
       "      <td>carnivora</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40103 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0   \\\n",
       "0                               \n",
       "1                               \n",
       "2                               \n",
       "3                               \n",
       "4                               \n",
       "...                       ...   \n",
       "40098           vulpes vulpes   \n",
       "40099           vulpes vulpes   \n",
       "40100           vulpes vulpes   \n",
       "40101           vulpes vulpes   \n",
       "40102  zalophus californianus   \n",
       "\n",
       "                                                      1       2       3   \\\n",
       "0                                                          10114   11118   \n",
       "1                                                          33554  693996   \n",
       "2                                                         371929  694002   \n",
       "3                                                          49442  694002   \n",
       "4                                                           9397   11118   \n",
       "...                                                  ...     ...     ...   \n",
       "40098                                 alphacoronavirus 1    9627  693997   \n",
       "40099  severe acute respiratory syndrome-related coro...    9627  694009   \n",
       "40100  severe acute respiratory syndrome-related coro...    9627  694009   \n",
       "40101  severe acute respiratory syndrome-related coro...    9627  694009   \n",
       "40102  severe acute respiratory syndrome-related coro...    9704  694009   \n",
       "\n",
       "         4     5      6            7              8              9   ...  \\\n",
       "0      TRUE  TRUE  FALSE       rattus        muridae       rodentia  ...   \n",
       "1      TRUE  TRUE  FALSE                                  carnivora  ...   \n",
       "2      TRUE  TRUE  FALSE     pitangus     tyrannidae  passeriformes  ...   \n",
       "3      TRUE  TRUE  FALSE  rhinolophus  rhinolophidae     chiroptera  ...   \n",
       "4      TRUE  TRUE  FALSE                                 chiroptera  ...   \n",
       "...     ...   ...    ...          ...            ...            ...  ...   \n",
       "40098  TRUE  TRUE   TRUE       vulpes        canidae      carnivora  ...   \n",
       "40099  TRUE  TRUE   TRUE       vulpes        canidae      carnivora  ...   \n",
       "40100  TRUE  TRUE   TRUE       vulpes        canidae      carnivora  ...   \n",
       "40101  TRUE  TRUE   TRUE       vulpes        canidae      carnivora  ...   \n",
       "40102  TRUE  TRUE   TRUE     zalophus      otariidae      carnivora  ...   \n",
       "\n",
       "             24 25 26 27 28 29 30      31                                 32  \\\n",
       "0                                                                              \n",
       "1                                                                              \n",
       "2                                                                              \n",
       "3                                                                              \n",
       "4                                                                              \n",
       "...         ... .. .. .. .. .. ..     ...                                ...   \n",
       "40098                                                                          \n",
       "40099  15921605                    219444  https://doi.org/10.1111/mec.15463   \n",
       "40100                                                                          \n",
       "40101                                                                          \n",
       "40102                                                                          \n",
       "\n",
       "      33  \n",
       "0         \n",
       "1         \n",
       "2         \n",
       "3         \n",
       "4         \n",
       "...   ..  \n",
       "40098     \n",
       "40099     \n",
       "40100     \n",
       "40101     \n",
       "40102     \n",
       "\n",
       "[40103 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open('/home/bio_kang/virus_host_virion/Virion.csv','rb') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "data_procession = []\n",
    "for line in lines:\n",
    "    if len(line) < 34:\n",
    "        continue\n",
    "    else:\n",
    "        data = line.decode('ISO-8859-1').strip('\\n').split('\\t')\n",
    "\n",
    "        if data[3] == '' or data[4] == \"FALSE\" or data[4] == '':\n",
    "            continue\n",
    "        else:\n",
    "            data_procession.append(data)\n",
    "\n",
    "\n",
    "\n",
    "data = pd.DataFrame(data_procession[1:-1],columns=data_procession[0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "getting taxon information and taxid of virus and host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "virus_ID = data[['VirusTaxID']]\n",
    "Host_ID = data[['HostTaxID']]\n",
    "\n",
    "virus_ID.to_csv('Virus_ID.csv',index=None,header=None)\n",
    "Host_ID.to_csv('Host_ID.csv',index=None,header=None)\n",
    "\n",
    "os.system('sort Virus_ID.csv | uniq | /home/bio_kang/software/anaconda3/envs/bio/bin/taxonkit reformat -I 1 -f \"{p};{c};{o};{f};{g};{s}\" -F -t > Virus_taxid.txt')\n",
    "os.system('sort Host_ID.csv | uniq | /home/bio_kang/software/anaconda3/envs/bio/bin/taxonkit reformat -I 1 -f \"{p};{c};{o};{f};{g};{s}\" -F -t > Host_taxid.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000586\tNegarnaviricota;Ellioviricetes;Bunyavirales;Hantaviridae;Orthohantavirus;Carrizal virus\t2497569;2497576;1980410;1980413;1980442;1000586\n",
      "1000587\tNegarnaviricota;Ellioviricetes;Bunyavirales;Hantaviridae;Orthohantavirus;Huitzilac virus\t2497569;2497576;1980410;1980413;1980442;1000587\n",
      "1000650\tPeploviricota;Herviviricetes;Herpesvirales;Herpesviridae;unclassified Herpesviridae genus;Arvicola amphibius gammaherpesvirus 1\t2731361;2731363;548681;10292;;1000650\n",
      "1001081\tNegarnaviricota;Ellioviricetes;Bunyavirales;Hantaviridae;Orthohantavirus;Azagny virus\t2497569;2497576;1980410;1980413;1980442;1001081\n",
      "1001627\tPisuviricota;Pisoniviricetes;Picornavirales;Caliciviridae;unclassified Caliciviridae genus;Calicivirus chicken/V0013/Bayern/2004\t2732408;2732506;464095;11974;;1001627\n",
      "1001628\tPisuviricota;Pisoniviricetes;Picornavirales;Caliciviridae;unclassified Caliciviridae genus;Calicivirus chicken/V0027/Bayern/2004\t2732408;2732506;464095;11974;;1001628\n",
      "1001974\tNegarnaviricota;Ellioviricetes;Bunyavirales;Hantaviridae;Orthohantavirus;Eothenomys miletus hantavirus LX17\t2497569;2497576;1980410;1980413;1980442;1001974\n",
      "1001975\tNegarnaviricota;Ellioviricetes;Bunyavirales;Hantaviridae;Orthohantavirus;Eothenomys miletus hantavirus LX43\t2497569;2497576;1980410;1980413;1980442;1001975\n",
      "100217\tNucleocytoviricota;Megaviricetes;Pimascovirales;Iridoviridae;Ranavirus;Epizootic haematopoietic necrosis virus\t2732007;2732523;2732555;10486;10492;100217\n",
      "1002931\tKitrinoviricota;Flasuviricetes;Amarillovirales;Flaviviridae;unclassified Flaviviridae genus;Bat GB-like virus JFD-2011\t2732406;2732462;2732545;11050;;1002931\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('head Virus_taxid.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "obtaining whole data I needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Virus_taxid.txt','r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "virus_taxon_info = {}\n",
    "virus_taxon_taxid_info = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip('\\n').split('\\t')\n",
    "    taxon = line[1].split(';')\n",
    "    taxid = line[2].split(';')\n",
    "\n",
    "    if len(taxon) !=6:\n",
    "        continue\n",
    "    else:\n",
    "        virus_taxon_info[line[0]] = taxon\n",
    "    if len(taxid) != 6:\n",
    "        continue\n",
    "    else:\n",
    "        virus_taxon_taxid_info[line[0]] = taxid\n",
    "    \n",
    "# write file\n",
    "write_taxon = pd.DataFrame(virus_taxon_info).T.reset_index()\n",
    "write_taxon.columns = ['Virus_ID','Phylum','Class','Order','Family','Genus','Species']\n",
    "write_taxon.to_csv('/home/bio_kang/Learning/graduation/Virus_taxon.csv',index=False)\n",
    "\n",
    "write_taxid = pd.DataFrame(virus_taxon_taxid_info).T.reset_index()\n",
    "write_taxid.columns = ['Virus_ID','Phylum_ID','Class_ID','Order_ID','Family_ID','Genus_ID','Species_ID']\n",
    "write_taxid.to_csv('/home/bio_kang/Learning/graduation/Virus_taxid.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open('Host_taxid.txt','r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "host_taxon_info = {}\n",
    "host_taxon_taxid_info = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip('\\n').split('\\t')\n",
    "    taxon = line[1].split(';')\n",
    "    taxid = line[2].split(';')\n",
    "\n",
    "    if len(taxon) !=6:\n",
    "        continue\n",
    "    else:\n",
    "        host_taxon_info[line[0]] = taxon\n",
    "    if len(taxid) != 6:\n",
    "        continue\n",
    "    else:\n",
    "        host_taxon_taxid_info[line[0]] = taxid\n",
    "    \n",
    "# write file\n",
    "write_taxon = pd.DataFrame(host_taxon_info).T.reset_index()\n",
    "write_taxon.columns = ['Host_ID','Phylum','Class','Order','Family','Genus','Species']\n",
    "write_taxon.to_csv('/home/bio_kang/Learning/graduation/Host_taxon.csv',index=False)\n",
    "\n",
    "write_taxid = pd.DataFrame(host_taxon_taxid_info).T.reset_index()\n",
    "write_taxid.columns = ['Host_ID','Phylum_ID','Class_ID','Order_ID','Family_ID','Genus_ID','Species_ID']\n",
    "write_taxid.to_csv('/home/bio_kang/Learning/graduation/Host_taxid.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "extracing coronaviridae data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open('/home/bio_kang/virus_host_virion/Virion.csv','rb') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "coronaviridae_data = []\n",
    "\n",
    "alphacoronavirus = []\n",
    "betacoronavirus = []\n",
    "deltacoronavirus =[]\n",
    "gammacoronavirus =[]\n",
    "\n",
    "header = lines[0].decode('ISO-8859-1').strip('\\n').split('\\t')\n",
    "\n",
    "for line in lines:\n",
    "    line = line.decode('ISO-8859-1').strip('\\n').split('\\t')\n",
    "    if len(line) < 34:\n",
    "        continue\n",
    "    else:\n",
    "        if line[13] == 'coronaviridae':\n",
    "            coronaviridae_data.append(line)\n",
    "\n",
    "            if line[12] == 'alphacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    alphacoronavirus.append(line)\n",
    "            elif line[12] == 'betacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    betacoronavirus.append(line)\n",
    "            elif line[12] == 'deltacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    deltacoronavirus.append(line)\n",
    "            elif line[12] == 'gammacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    gammacoronavirus.append(line)\n",
    "\n",
    "data = pd.DataFrame(coronaviridae_data, columns=header)\n",
    "\n",
    "alpha = pd.DataFrame(alphacoronavirus, columns=header)\n",
    "beta = pd.DataFrame(betacoronavirus, columns=header)\n",
    "delta = pd.DataFrame(deltacoronavirus, columns=header)\n",
    "gamma = pd.DataFrame(gammacoronavirus, columns=header)\n",
    "\n",
    "alpha.to_csv('alphacoronavirus.csv',index=False, sep='\\t')\n",
    "beta.to_csv('betacoronavirus.csv',index=False, sep='\\t')\n",
    "delta.to_csv('deltacoronavirus.csv',index=False, sep='\\t')\n",
    "gamma.to_csv('gammacoronavirus.csv',index=False, sep='\\t')\n",
    "\n",
    "data.to_csv('coronaviridae_virus.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "\n",
    "os.system(\"awk -F '\\t' '{print $34}' alphacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > alphacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' betacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > betacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' gammacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > gammacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' deltacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > deltacoronavirus.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "genus_set = [\"alphacoronavirus\",\"betacoronavirus\",\"gammacoronavirus\",\"deltacoronavirus\"]\n",
    "\n",
    "for genus in genus_set:\n",
    "    os.system(\"grep -P '[A-Z]\\w+\\.\\d' {}.txt | sort -u > criterion_format_{}.txt\".format(genus,genus))\n",
    "    os.system(\"grep -P '[A-Z]\\w+$' {}.txt | sed 's/^\\s//' | sort -u > non_version_{}.txt\".format(genus,genus))\n",
    "    os.system(\"grep -P '^\\d+' {}.txt | sort -u > pure_num_{}.txt\".format(genus,genus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "import os\n",
    "\n",
    "def from_accession_to_id(segment):\n",
    "    term_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&api_key=6e2f58038b074d964bb8b87f43623af7b508&term=' + \",\".join(segment)\n",
    "    term_info = requests.get(term_url, timeout=(5,5)).text\n",
    " \n",
    "    id_pattern = re.compile(r'<Id>(\\d+)</Id>')\n",
    "    accession_id = re.findall(id_pattern, term_info)\n",
    "    return accession_id\n",
    "\n",
    "def from_id_to_version_accession(id):\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text\n",
    "\n",
    "    version_pattern = re.compile(r'>[A-Z]{1,2}\\d+.\\d')\n",
    "    version_info = re.findall(version_pattern, fasta_info)\n",
    "    return version_info\n",
    "\n",
    "\n",
    "genus_set = [\"alphacoronavirus\",\"betacoronavirus\",\"gammacoronavirus\",\"deltacoronavirus\"]\n",
    "\n",
    "for genus in genus_set:\n",
    "    with open('non_version_{}.txt'.format(genus), 'r') as f:\n",
    "        f_lines = f.readlines()\n",
    "    with open('pure_num_{}.txt'.format(genus),'r') as k:\n",
    "        k_lines = k.readlines()\n",
    "\n",
    "    f_data = list(map(lambda x:x.strip('\\n'), f_lines))\n",
    "    f_time = len(f_data) // 20 + 1\n",
    "\n",
    "    k_data = list(map(lambda x:x.strip('\\n'),k_lines))\n",
    "    k_time = len(k_data) // 20 +1\n",
    "\n",
    "    version = []\n",
    "    \n",
    "    for i in range(f_time):\n",
    "        segment = f_data[i*20:(i+1)*20]\n",
    "        id = from_accession_to_id(segment)\n",
    "        version = version + from_id_to_version_accession(id)\n",
    "\n",
    "    for j in range(k_time):\n",
    "        id = k_data[i*20:(i+1)*20]\n",
    "        version = version + from_id_to_version_accession(id)\n",
    "\n",
    "    with open('including_version_{}.txt'.format(genus), 'w') as m:\n",
    "        m.writelines(list(map(lambda x:x+'\\n',version)))\n",
    "\n",
    "    os.system(\"grep -oP '[A-Z]+\\d+.\\d' including_version_{}.txt >> criterion_format_{}.txt\".format(genus,genus))\n",
    "    os.system(\"sort -u criterion_format_{}.txt > finally_{}.txt\".format(genus,genus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_id_to_version_accession(id):\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text\n",
    "\n",
    "    version_pattern = re.compile(r'>[A-Z]{1,2}\\d+.\\d')\n",
    "    version_info = re.findall(version_pattern, fasta_info)\n",
    "    return version_info\n",
    "\n",
    "with open('pure_num_alphacoronavirus.txt', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "data = list(map(lambda x:x.strip('\\n'),lines))\n",
    "time = len(data) // 20 + 1\n",
    "for i in range(time):\n",
    "    id = data[i*20:(i+1)*20]\n",
    "    print(from_id_to_version_accession(id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "\n",
    "def get_url_info_id(segment,num):\n",
    "    # Using requests function get the id from term\n",
    "    term_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&api_key=6e2f58038b074d964bb8b87f43623af7b508&term=' + \",\".join(segment)\n",
    "    term_info = requests.get(term_url, timeout=(5,5)).text\n",
    " \n",
    "    id_pattern = re.compile(r'<Id>(\\d+)</Id>')\n",
    "    accession_id = re.findall(id_pattern, term_info)\n",
    "\n",
    "    # get sequencing information from id\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(accession_id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e4c0b1c8e35d9f80793da27c1f09cec2906743a08f4271636815a29bc833a021"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
