{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data process: Remove records of virus or host withou accession number of NCBI \n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "data_procession = []\n",
    "with open('/home/ouconstand/data/Virus_Host/Virion.csv','r', encoding='ISO-8859-1') as f:\n",
    "    for line in f.read().splitlines():\n",
    "        data = line.split('\\t')\n",
    "        if len(data) < 34:\n",
    "            continue\n",
    "        else:\n",
    "            #print(len(data))\n",
    "            if data[3] == '' or data[4] == \"FALSE\" or data[4] == '' or data[33] == '':\n",
    "                continue\n",
    "            else:\n",
    "                data_procession.append(data)\n",
    "\n",
    "data = pd.DataFrame(data_procession[1:-1],columns=data_procession[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Host</th>\n",
       "      <th>Virus</th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>VirusTaxID</th>\n",
       "      <th>HostNCBIResolved</th>\n",
       "      <th>VirusNCBIResolved</th>\n",
       "      <th>ICTVRatified</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>...</th>\n",
       "      <th>PMID</th>\n",
       "      <th>ReleaseYear</th>\n",
       "      <th>ReleaseMonth</th>\n",
       "      <th>ReleaseDay</th>\n",
       "      <th>CollectionYear</th>\n",
       "      <th>CollectionMonth</th>\n",
       "      <th>CollectionDay</th>\n",
       "      <th>AssocID</th>\n",
       "      <th>DatabaseDOI</th>\n",
       "      <th>NCBIAccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>aalivirus a</td>\n",
       "      <td>8830</td>\n",
       "      <td>2169685</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td></td>\n",
       "      <td>anatidae</td>\n",
       "      <td>anseriformes</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2021</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>2018</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MT250552.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>adeno-associated virus</td>\n",
       "      <td>9443</td>\n",
       "      <td>272636</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primates</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2014</td>\n",
       "      <td>7</td>\n",
       "      <td>28</td>\n",
       "      <td>2014</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KJ743606.1, KJ743607.1, KJ743608.1, KJ743609.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>adeno-associated virus</td>\n",
       "      <td>9443</td>\n",
       "      <td>272636</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primates</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2014</td>\n",
       "      <td>7</td>\n",
       "      <td>28</td>\n",
       "      <td>2014</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KJ743603.1, KJ743604.1, KJ743605.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>adeno-associated virus</td>\n",
       "      <td>9443</td>\n",
       "      <td>272636</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primates</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2017</td>\n",
       "      <td>12</td>\n",
       "      <td>31</td>\n",
       "      <td>2015</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KX602251.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td>adenoviridae sp.</td>\n",
       "      <td>8487</td>\n",
       "      <td>2558248</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>testudinidae</td>\n",
       "      <td>testudines</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2021</td>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "      <td>2014</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MT900848.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428004</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>2011</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KF481954.1, KF494345.1, KF735065.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428005</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2019</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>2011</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NC_043407.1, NC_043408.1, NC_043409.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428006</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>pirital mammarenavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>49891</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2012</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1994</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>JF412368.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428007</th>\n",
       "      <td>zygodontomys cherriei</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>218840</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2012</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>2008</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>JN717148.1, JN717149.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428008</th>\n",
       "      <td>zygodontomys cherriei</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>218840</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>2008</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KM196124.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>428009 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Host                    Virus HostTaxID VirusTaxID  \\\n",
       "0                                            aalivirus a      8830    2169685   \n",
       "1                                 adeno-associated virus      9443     272636   \n",
       "2                                 adeno-associated virus      9443     272636   \n",
       "3                                 adeno-associated virus      9443     272636   \n",
       "4                                       adenoviridae sp.      8487    2558248   \n",
       "...                         ...                      ...       ...        ...   \n",
       "428004  zygodontomys brevicauda  necocli orthohantavirus    157541    1980482   \n",
       "428005  zygodontomys brevicauda  necocli orthohantavirus    157541    1980482   \n",
       "428006  zygodontomys brevicauda   pirital mammarenavirus    157541      49891   \n",
       "428007    zygodontomys cherriei  necocli orthohantavirus    218840    1980482   \n",
       "428008    zygodontomys cherriei  necocli orthohantavirus    218840    1980482   \n",
       "\n",
       "       HostNCBIResolved VirusNCBIResolved ICTVRatified     HostGenus  \\\n",
       "0                  TRUE              TRUE         TRUE                 \n",
       "1                  TRUE              TRUE        FALSE                 \n",
       "2                  TRUE              TRUE        FALSE                 \n",
       "3                  TRUE              TRUE        FALSE                 \n",
       "4                  TRUE              TRUE        FALSE                 \n",
       "...                 ...               ...          ...           ...   \n",
       "428004             TRUE              TRUE         TRUE  zygodontomys   \n",
       "428005             TRUE              TRUE         TRUE  zygodontomys   \n",
       "428006             TRUE              TRUE         TRUE  zygodontomys   \n",
       "428007             TRUE              TRUE         TRUE  zygodontomys   \n",
       "428008             TRUE              TRUE         TRUE  zygodontomys   \n",
       "\n",
       "          HostFamily     HostOrder  ... PMID ReleaseYear ReleaseMonth  \\\n",
       "0           anatidae  anseriformes  ...             2021            3   \n",
       "1                         primates  ...             2014            7   \n",
       "2                         primates  ...             2014            7   \n",
       "3                         primates  ...             2017           12   \n",
       "4       testudinidae    testudines  ...             2021            5   \n",
       "...              ...           ...  ...  ...         ...          ...   \n",
       "428004    cricetidae      rodentia  ...             2015            8   \n",
       "428005    cricetidae      rodentia  ...             2019            6   \n",
       "428006    cricetidae      rodentia  ...             2012            5   \n",
       "428007    cricetidae      rodentia  ...             2012            2   \n",
       "428008    cricetidae      rodentia  ...             2015            8   \n",
       "\n",
       "       ReleaseDay CollectionYear CollectionMonth CollectionDay AssocID  \\\n",
       "0               8           2018                                         \n",
       "1              28           2014               3            10           \n",
       "2              28           2014               3             5           \n",
       "3              31           2015              12            12           \n",
       "4              16           2014              12                         \n",
       "...           ...            ...             ...           ...     ...   \n",
       "428004         29           2011               2             7           \n",
       "428005         28           2011               2             7           \n",
       "428006          7           1994                                         \n",
       "428007         15           2008                                         \n",
       "428008         29           2008               3            16           \n",
       "\n",
       "       DatabaseDOI                                   NCBIAccession  \n",
       "0                                                       MT250552.1  \n",
       "1                   KJ743606.1, KJ743607.1, KJ743608.1, KJ743609.1  \n",
       "2                               KJ743603.1, KJ743604.1, KJ743605.1  \n",
       "3                                                       KX602251.1  \n",
       "4                                                       MT900848.1  \n",
       "...            ...                                             ...  \n",
       "428004                          KF481954.1, KF494345.1, KF735065.1  \n",
       "428005                       NC_043407.1, NC_043408.1, NC_043409.1  \n",
       "428006                                                  JF412368.1  \n",
       "428007                                      JN717148.1, JN717149.1  \n",
       "428008                                                  KM196124.1  \n",
       "\n",
       "[428009 rows x 34 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Host</th>\n",
       "      <th>Virus</th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>VirusTaxID</th>\n",
       "      <th>HostNCBIResolved</th>\n",
       "      <th>VirusNCBIResolved</th>\n",
       "      <th>ICTVRatified</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>...</th>\n",
       "      <th>PMID</th>\n",
       "      <th>ReleaseYear</th>\n",
       "      <th>ReleaseMonth</th>\n",
       "      <th>ReleaseDay</th>\n",
       "      <th>CollectionYear</th>\n",
       "      <th>CollectionMonth</th>\n",
       "      <th>CollectionDay</th>\n",
       "      <th>AssocID</th>\n",
       "      <th>DatabaseDOI</th>\n",
       "      <th>NCBIAccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td></td>\n",
       "      <td>adenovirus predict_adv-13</td>\n",
       "      <td>27659</td>\n",
       "      <td>2710946</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>sturnira</td>\n",
       "      <td>phyllostomidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2020</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2011</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MK325782.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td></td>\n",
       "      <td>adenovirus predict_adv-13</td>\n",
       "      <td>27659</td>\n",
       "      <td>2710946</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>sturnira</td>\n",
       "      <td>phyllostomidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2021</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>2011</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MK325782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td></td>\n",
       "      <td>adenovirus predict_adv-19</td>\n",
       "      <td>10111</td>\n",
       "      <td>2710952</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>praomys</td>\n",
       "      <td>muridae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2020</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MK325792.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td></td>\n",
       "      <td>adenovirus predict_adv-19</td>\n",
       "      <td>10111</td>\n",
       "      <td>2710952</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>praomys</td>\n",
       "      <td>muridae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2021</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MK325792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td></td>\n",
       "      <td>adenovirus predict_adv-21</td>\n",
       "      <td>10111</td>\n",
       "      <td>2710955</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>praomys</td>\n",
       "      <td>muridae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2020</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2010</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MK325833.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428004</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>2011</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KF481954.1, KF494345.1, KF735065.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428005</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2019</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>2011</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NC_043407.1, NC_043408.1, NC_043409.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428006</th>\n",
       "      <td>zygodontomys brevicauda</td>\n",
       "      <td>pirital mammarenavirus</td>\n",
       "      <td>157541</td>\n",
       "      <td>49891</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2012</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1994</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>JF412368.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428007</th>\n",
       "      <td>zygodontomys cherriei</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>218840</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2012</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>2008</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>JN717148.1, JN717149.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428008</th>\n",
       "      <td>zygodontomys cherriei</td>\n",
       "      <td>necocli orthohantavirus</td>\n",
       "      <td>218840</td>\n",
       "      <td>1980482</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>zygodontomys</td>\n",
       "      <td>cricetidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>2008</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>KM196124.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>420174 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Host                      Virus HostTaxID  \\\n",
       "6                                adenovirus predict_adv-13     27659   \n",
       "7                                adenovirus predict_adv-13     27659   \n",
       "13                               adenovirus predict_adv-19     10111   \n",
       "14                               adenovirus predict_adv-19     10111   \n",
       "15                               adenovirus predict_adv-21     10111   \n",
       "...                         ...                        ...       ...   \n",
       "428004  zygodontomys brevicauda    necocli orthohantavirus    157541   \n",
       "428005  zygodontomys brevicauda    necocli orthohantavirus    157541   \n",
       "428006  zygodontomys brevicauda     pirital mammarenavirus    157541   \n",
       "428007    zygodontomys cherriei    necocli orthohantavirus    218840   \n",
       "428008    zygodontomys cherriei    necocli orthohantavirus    218840   \n",
       "\n",
       "       VirusTaxID HostNCBIResolved VirusNCBIResolved ICTVRatified  \\\n",
       "6         2710946             TRUE              TRUE        FALSE   \n",
       "7         2710946             TRUE              TRUE        FALSE   \n",
       "13        2710952             TRUE              TRUE        FALSE   \n",
       "14        2710952             TRUE              TRUE        FALSE   \n",
       "15        2710955             TRUE              TRUE        FALSE   \n",
       "...           ...              ...               ...          ...   \n",
       "428004    1980482             TRUE              TRUE         TRUE   \n",
       "428005    1980482             TRUE              TRUE         TRUE   \n",
       "428006      49891             TRUE              TRUE         TRUE   \n",
       "428007    1980482             TRUE              TRUE         TRUE   \n",
       "428008    1980482             TRUE              TRUE         TRUE   \n",
       "\n",
       "           HostGenus      HostFamily   HostOrder  ... PMID ReleaseYear  \\\n",
       "6           sturnira  phyllostomidae  chiroptera  ...             2020   \n",
       "7           sturnira  phyllostomidae  chiroptera  ...             2021   \n",
       "13           praomys         muridae    rodentia  ...             2020   \n",
       "14           praomys         muridae    rodentia  ...             2021   \n",
       "15           praomys         muridae    rodentia  ...             2020   \n",
       "...              ...             ...         ...  ...  ...         ...   \n",
       "428004  zygodontomys      cricetidae    rodentia  ...             2015   \n",
       "428005  zygodontomys      cricetidae    rodentia  ...             2019   \n",
       "428006  zygodontomys      cricetidae    rodentia  ...             2012   \n",
       "428007  zygodontomys      cricetidae    rodentia  ...             2012   \n",
       "428008  zygodontomys      cricetidae    rodentia  ...             2015   \n",
       "\n",
       "       ReleaseMonth ReleaseDay CollectionYear CollectionMonth CollectionDay  \\\n",
       "6                 3          1           2011              11            20   \n",
       "7                 3          9           2011              11             1   \n",
       "13                3          1           2009               4             1   \n",
       "14                3          9           2009               4             1   \n",
       "15                3          1           2010               1             5   \n",
       "...             ...        ...            ...             ...           ...   \n",
       "428004            8         29           2011               2             7   \n",
       "428005            6         28           2011               2             7   \n",
       "428006            5          7           1994                                 \n",
       "428007            2         15           2008                                 \n",
       "428008            8         29           2008               3            16   \n",
       "\n",
       "       AssocID DatabaseDOI                          NCBIAccession  \n",
       "6                                                      MK325782.1  \n",
       "7                                                        MK325782  \n",
       "13                                                     MK325792.1  \n",
       "14                                                       MK325792  \n",
       "15                                                     MK325833.1  \n",
       "...        ...         ...                                    ...  \n",
       "428004                         KF481954.1, KF494345.1, KF735065.1  \n",
       "428005                      NC_043407.1, NC_043408.1, NC_043409.1  \n",
       "428006                                                 JF412368.1  \n",
       "428007                                     JN717148.1, JN717149.1  \n",
       "428008                                                 KM196124.1  \n",
       "\n",
       "[420174 rows x 34 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data filtering, without Genus level\n",
    "with_genus_data = data[data['HostGenus'] != '']\n",
    "with_genus_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conserve the data with Genus level taxon and NCBIAccession\n",
    "with_genus_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/data_with_genus_accession.csv',index=None)\n",
    "# and the file is moved to /home/ouconstand/data/Virus_Host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract coroviridae data and conserve it\n",
    "\"\"\"\n",
    "coro_data = with_genus_data[with_genus_data['VirusFamily']=='coronaviridae']  \n",
    "coro_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_data.csv',index=None)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(35611, 34)\n",
      "(1544, 34)\n"
     ]
    }
   ],
   "source": [
    "# remove duplication\n",
    "coro_data = pd.read_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/coro_data.csv')\n",
    "print(coro_data.shape)\n",
    "coro_data.drop_duplicates(subset=['HostTaxID', 'VirusTaxID'], inplace=True)\n",
    "print(coro_data.shape)\n",
    "coro_data.to_csv('/home/ouconstand/Learning/graduation/virus_host/construct_graph/remove_duplication_coro_data.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Host</th>\n",
       "      <th>Virus</th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>VirusTaxID</th>\n",
       "      <th>HostNCBIResolved</th>\n",
       "      <th>VirusNCBIResolved</th>\n",
       "      <th>ICTVRatified</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>...</th>\n",
       "      <th>PMID</th>\n",
       "      <th>ReleaseYear</th>\n",
       "      <th>ReleaseMonth</th>\n",
       "      <th>ReleaseDay</th>\n",
       "      <th>CollectionYear</th>\n",
       "      <th>CollectionMonth</th>\n",
       "      <th>CollectionDay</th>\n",
       "      <th>AssocID</th>\n",
       "      <th>DatabaseDOI</th>\n",
       "      <th>NCBIAccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus btcov/ncl_mco1/mop_con/rsa/2012</td>\n",
       "      <td>258862</td>\n",
       "      <td>1532174</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>mops</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KF843853.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus p.sp/k/spain/2007</td>\n",
       "      <td>27671</td>\n",
       "      <td>1076425</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HQ184060.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus sp.</td>\n",
       "      <td>27671</td>\n",
       "      <td>1906673</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>11.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LC334391.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>alphacoronavirus tadaridam4</td>\n",
       "      <td>9437</td>\n",
       "      <td>1582306</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>tadarida</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KM514667.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>avian coronavirus</td>\n",
       "      <td>53274</td>\n",
       "      <td>694014</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>rhizomys</td>\n",
       "      <td>spalacidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021</td>\n",
       "      <td>6.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT222079.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35600</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>alpaca coronavirus</td>\n",
       "      <td>30538</td>\n",
       "      <td>1770168</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015</td>\n",
       "      <td>12.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KP893231.1, KP893232.1, KP893233.1, KP893234.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35604</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>alpaca coronavirus ca08-1/2008</td>\n",
       "      <td>30538</td>\n",
       "      <td>669204</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GQ398771.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35605</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>avian coronavirus</td>\n",
       "      <td>30538</td>\n",
       "      <td>694014</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>218959.0</td>\n",
       "      <td>https://doi.org/10.1038/sdata.2015.49</td>\n",
       "      <td>372468322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35608</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>betacoronavirus 1</td>\n",
       "      <td>30538</td>\n",
       "      <td>694003</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>218960.0</td>\n",
       "      <td>https://doi.org/10.1038/sdata.2015.49</td>\n",
       "      <td>124294673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35610</th>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>human coronavirus 229e</td>\n",
       "      <td>30538</td>\n",
       "      <td>11137</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2012</td>\n",
       "      <td>4.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JQ410000.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1544 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Host                                             Virus  \\\n",
       "0                NaN  alphacoronavirus btcov/ncl_mco1/mop_con/rsa/2012   \n",
       "1                NaN                alphacoronavirus p.sp/k/spain/2007   \n",
       "2                NaN                              alphacoronavirus sp.   \n",
       "3                NaN                       alphacoronavirus tadaridam4   \n",
       "4                NaN                                 avian coronavirus   \n",
       "...              ...                                               ...   \n",
       "35600  vicugna pacos                                alpaca coronavirus   \n",
       "35604  vicugna pacos                    alpaca coronavirus ca08-1/2008   \n",
       "35605  vicugna pacos                                 avian coronavirus   \n",
       "35608  vicugna pacos                                 betacoronavirus 1   \n",
       "35610  vicugna pacos                            human coronavirus 229e   \n",
       "\n",
       "       HostTaxID  VirusTaxID  HostNCBIResolved  VirusNCBIResolved  \\\n",
       "0         258862     1532174              True               True   \n",
       "1          27671     1076425              True               True   \n",
       "2          27671     1906673              True               True   \n",
       "3           9437     1582306              True               True   \n",
       "4          53274      694014              True               True   \n",
       "...          ...         ...               ...                ...   \n",
       "35600      30538     1770168              True               True   \n",
       "35604      30538      669204              True               True   \n",
       "35605      30538      694014              True               True   \n",
       "35608      30538      694003              True               True   \n",
       "35610      30538       11137              True               True   \n",
       "\n",
       "       ICTVRatified     HostGenus        HostFamily     HostOrder  ... PMID  \\\n",
       "0             False          mops        molossidae    chiroptera  ...  NaN   \n",
       "1             False  pipistrellus  vespertilionidae    chiroptera  ...  NaN   \n",
       "2             False  pipistrellus  vespertilionidae    chiroptera  ...  NaN   \n",
       "3             False      tadarida        molossidae    chiroptera  ...  NaN   \n",
       "4              True      rhizomys        spalacidae      rodentia  ...  NaN   \n",
       "...             ...           ...               ...           ...  ...  ...   \n",
       "35600         False       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35604         False       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35605          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35608          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "35610          True       vicugna         camelidae  artiodactyla  ...  NaN   \n",
       "\n",
       "      ReleaseYear ReleaseMonth ReleaseDay CollectionYear CollectionMonth  \\\n",
       "0            2014         12.0       12.0         2012.0             NaN   \n",
       "1            2011          8.0       21.0         2007.0             NaN   \n",
       "2            2017         11.0       28.0            NaN             NaN   \n",
       "3            2014         12.0       17.0         2011.0             NaN   \n",
       "4            2021          6.0       23.0         2014.0             3.0   \n",
       "...           ...          ...        ...            ...             ...   \n",
       "35600        2015         12.0       20.0         2009.0             NaN   \n",
       "35604        2009          9.0        1.0         2008.0             NaN   \n",
       "35605        2005          NaN        NaN            NaN             NaN   \n",
       "35608        2006          NaN        NaN            NaN             NaN   \n",
       "35610        2012          4.0       29.0         2008.0             NaN   \n",
       "\n",
       "      CollectionDay   AssocID                            DatabaseDOI  \\\n",
       "0               NaN       NaN                                    NaN   \n",
       "1               NaN       NaN                                    NaN   \n",
       "2               NaN       NaN                                    NaN   \n",
       "3               NaN       NaN                                    NaN   \n",
       "4              19.0       NaN                                    NaN   \n",
       "...             ...       ...                                    ...   \n",
       "35600           NaN       NaN                                    NaN   \n",
       "35604           NaN       NaN                                    NaN   \n",
       "35605           NaN  218959.0  https://doi.org/10.1038/sdata.2015.49   \n",
       "35608           NaN  218960.0  https://doi.org/10.1038/sdata.2015.49   \n",
       "35610           NaN       NaN                                    NaN   \n",
       "\n",
       "                                           NCBIAccession  \n",
       "0                                             KF843853.1  \n",
       "1                                             HQ184060.1  \n",
       "2                                             LC334391.1  \n",
       "3                                             KM514667.1  \n",
       "4                                             MT222079.1  \n",
       "...                                                  ...  \n",
       "35600  KP893231.1, KP893232.1, KP893233.1, KP893234.1...  \n",
       "35604                                         GQ398771.1  \n",
       "35605                                          372468322  \n",
       "35608                                          124294673  \n",
       "35610                                         JQ410000.1  \n",
       "\n",
       "[1544 rows x 34 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coro_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "下面的暂时先不用管（11 Jan）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "coro_taxon_host_data = coro_data[['HostTaxID', 'Host', 'HostGenus', 'HostFamily', 'HostOrder', 'HostClass']]\n",
    "coro_taxon_virus_data = coro_data[['VirusTaxID', 'Virus', 'VirusGenus', 'VirusFamily', 'VirusOrder', 'VirusClass']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HostTaxID</th>\n",
       "      <th>Host</th>\n",
       "      <th>HostGenus</th>\n",
       "      <th>HostFamily</th>\n",
       "      <th>HostOrder</th>\n",
       "      <th>HostClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>258862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mops</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pipistrellus</td>\n",
       "      <td>vespertilionidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9437</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tadarida</td>\n",
       "      <td>molossidae</td>\n",
       "      <td>chiroptera</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>53274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhizomys</td>\n",
       "      <td>spalacidae</td>\n",
       "      <td>rodentia</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35606</th>\n",
       "      <td>30538</td>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35607</th>\n",
       "      <td>30538</td>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35608</th>\n",
       "      <td>30538</td>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35609</th>\n",
       "      <td>30538</td>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35610</th>\n",
       "      <td>30538</td>\n",
       "      <td>vicugna pacos</td>\n",
       "      <td>vicugna</td>\n",
       "      <td>camelidae</td>\n",
       "      <td>artiodactyla</td>\n",
       "      <td>mammalia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35611 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       HostTaxID           Host     HostGenus        HostFamily     HostOrder  \\\n",
       "0         258862            NaN          mops        molossidae    chiroptera   \n",
       "1          27671            NaN  pipistrellus  vespertilionidae    chiroptera   \n",
       "2          27671            NaN  pipistrellus  vespertilionidae    chiroptera   \n",
       "3           9437            NaN      tadarida        molossidae    chiroptera   \n",
       "4          53274            NaN      rhizomys        spalacidae      rodentia   \n",
       "...          ...            ...           ...               ...           ...   \n",
       "35606      30538  vicugna pacos       vicugna         camelidae  artiodactyla   \n",
       "35607      30538  vicugna pacos       vicugna         camelidae  artiodactyla   \n",
       "35608      30538  vicugna pacos       vicugna         camelidae  artiodactyla   \n",
       "35609      30538  vicugna pacos       vicugna         camelidae  artiodactyla   \n",
       "35610      30538  vicugna pacos       vicugna         camelidae  artiodactyla   \n",
       "\n",
       "      HostClass  \n",
       "0      mammalia  \n",
       "1      mammalia  \n",
       "2      mammalia  \n",
       "3      mammalia  \n",
       "4      mammalia  \n",
       "...         ...  \n",
       "35606  mammalia  \n",
       "35607  mammalia  \n",
       "35608  mammalia  \n",
       "35609  mammalia  \n",
       "35610  mammalia  \n",
       "\n",
       "[35611 rows x 6 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coro_taxon_host_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Virus_taxid.txt','r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "virus_taxon_info = {}\n",
    "virus_taxon_taxid_info = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip('\\n').split('\\t')\n",
    "    taxon = line[1].split(';')\n",
    "    taxid = line[2].split(';')\n",
    "\n",
    "    if len(taxon) !=6:\n",
    "        continue\n",
    "    else:\n",
    "        virus_taxon_info[line[0]] = taxon\n",
    "    if len(taxid) != 6:\n",
    "        continue\n",
    "    else:\n",
    "        virus_taxon_taxid_info[line[0]] = taxid\n",
    "    \n",
    "# write file\n",
    "write_taxon = pd.DataFrame(virus_taxon_info).T.reset_index()\n",
    "write_taxon.columns = ['Virus_ID','Phylum','Class','Order','Family','Genus','Species']\n",
    "write_taxon.to_csv('/home/bio_kang/Learning/graduation/Virus_taxon.csv',index=False)\n",
    "\n",
    "write_taxid = pd.DataFrame(virus_taxon_taxid_info).T.reset_index()\n",
    "write_taxid.columns = ['Virus_ID','Phylum_ID','Class_ID','Order_ID','Family_ID','Genus_ID','Species_ID']\n",
    "write_taxid.to_csv('/home/bio_kang/Learning/graduation/Virus_taxid.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open('Host_taxid.txt','r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "host_taxon_info = {}\n",
    "host_taxon_taxid_info = {}\n",
    "\n",
    "for line in lines:\n",
    "    line = line.strip('\\n').split('\\t')\n",
    "    taxon = line[1].split(';')\n",
    "    taxid = line[2].split(';')\n",
    "\n",
    "    if len(taxon) !=6:\n",
    "        continue\n",
    "    else:\n",
    "        host_taxon_info[line[0]] = taxon\n",
    "    if len(taxid) != 6:\n",
    "        continue\n",
    "    else:\n",
    "        host_taxon_taxid_info[line[0]] = taxid\n",
    "    \n",
    "# write file\n",
    "write_taxon = pd.DataFrame(host_taxon_info).T.reset_index()\n",
    "write_taxon.columns = ['Host_ID','Phylum','Class','Order','Family','Genus','Species']\n",
    "write_taxon.to_csv('/home/bio_kang/Learning/graduation/Host_taxon.csv',index=False)\n",
    "\n",
    "write_taxid = pd.DataFrame(host_taxon_taxid_info).T.reset_index()\n",
    "write_taxid.columns = ['Host_ID','Phylum_ID','Class_ID','Order_ID','Family_ID','Genus_ID','Species_ID']\n",
    "write_taxid.to_csv('/home/bio_kang/Learning/graduation/Host_taxid.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "extracing coronaviridae data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "with open('/home/bio_kang/virus_host_virion/Virion.csv','rb') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "coronaviridae_data = []\n",
    "\n",
    "alphacoronavirus = []\n",
    "betacoronavirus = []\n",
    "deltacoronavirus =[]\n",
    "gammacoronavirus =[]\n",
    "\n",
    "header = lines[0].decode('ISO-8859-1').strip('\\n').split('\\t')\n",
    "\n",
    "for line in lines:\n",
    "    line = line.decode('ISO-8859-1').strip('\\n').split('\\t')\n",
    "    if len(line) < 34:\n",
    "        continue\n",
    "    else:\n",
    "        if line[13] == 'coronaviridae':\n",
    "            coronaviridae_data.append(line)\n",
    "\n",
    "            if line[12] == 'alphacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    alphacoronavirus.append(line)\n",
    "            elif line[12] == 'betacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    betacoronavirus.append(line)\n",
    "            elif line[12] == 'deltacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    deltacoronavirus.append(line)\n",
    "            elif line[12] == 'gammacoronavirus':\n",
    "                if line[33] != '':\n",
    "                    gammacoronavirus.append(line)\n",
    "\n",
    "data = pd.DataFrame(coronaviridae_data, columns=header)\n",
    "\n",
    "alpha = pd.DataFrame(alphacoronavirus, columns=header)\n",
    "beta = pd.DataFrame(betacoronavirus, columns=header)\n",
    "delta = pd.DataFrame(deltacoronavirus, columns=header)\n",
    "gamma = pd.DataFrame(gammacoronavirus, columns=header)\n",
    "\n",
    "alpha.to_csv('alphacoronavirus.csv',index=False, sep='\\t')\n",
    "beta.to_csv('betacoronavirus.csv',index=False, sep='\\t')\n",
    "delta.to_csv('deltacoronavirus.csv',index=False, sep='\\t')\n",
    "gamma.to_csv('gammacoronavirus.csv',index=False, sep='\\t')\n",
    "\n",
    "data.to_csv('coronaviridae_virus.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "\n",
    "os.system(\"awk -F '\\t' '{print $34}' alphacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > alphacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' betacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > betacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' gammacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > gammacoronavirus.txt\")\n",
    "os.system(\"awk -F '\\t' '{print $34}' deltacoronavirus.csv | sed 's/^,//' | awk -F ',' '{print $1}' | sed '1s/.*//' | sed '/^\\s*$/d' > deltacoronavirus.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "genus_set = [\"alphacoronavirus\",\"betacoronavirus\",\"gammacoronavirus\",\"deltacoronavirus\"]\n",
    "\n",
    "for genus in genus_set:\n",
    "    os.system(\"grep -P '[A-Z]\\w+\\.\\d' {}.txt | sort -u > criterion_format_{}.txt\".format(genus,genus))\n",
    "    os.system(\"grep -P '[A-Z]\\w+$' {}.txt | sed 's/^\\s//' | sort -u > non_version_{}.txt\".format(genus,genus))\n",
    "    os.system(\"grep -P '^\\d+' {}.txt | sort -u > pure_num_{}.txt\".format(genus,genus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "import os\n",
    "\n",
    "def from_accession_to_id(segment):\n",
    "    term_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&api_key=6e2f58038b074d964bb8b87f43623af7b508&term=' + \",\".join(segment)\n",
    "    term_info = requests.get(term_url, timeout=(5,5)).text\n",
    " \n",
    "    id_pattern = re.compile(r'<Id>(\\d+)</Id>')\n",
    "    accession_id = re.findall(id_pattern, term_info)\n",
    "    return accession_id\n",
    "\n",
    "def from_id_to_version_accession(id):\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text\n",
    "\n",
    "    version_pattern = re.compile(r'>[A-Z]{1,2}\\d+.\\d')\n",
    "    version_info = re.findall(version_pattern, fasta_info)\n",
    "    return version_info\n",
    "\n",
    "\n",
    "genus_set = [\"alphacoronavirus\",\"betacoronavirus\",\"gammacoronavirus\",\"deltacoronavirus\"]\n",
    "\n",
    "for genus in genus_set:\n",
    "    with open('non_version_{}.txt'.format(genus), 'r') as f:\n",
    "        f_lines = f.readlines()\n",
    "    with open('pure_num_{}.txt'.format(genus),'r') as k:\n",
    "        k_lines = k.readlines()\n",
    "\n",
    "    f_data = list(map(lambda x:x.strip('\\n'), f_lines))\n",
    "    f_time = len(f_data) // 20 + 1\n",
    "\n",
    "    k_data = list(map(lambda x:x.strip('\\n'),k_lines))\n",
    "    k_time = len(k_data) // 20 +1\n",
    "\n",
    "    version = []\n",
    "    \n",
    "    for i in range(f_time):\n",
    "        segment = f_data[i*20:(i+1)*20]\n",
    "        id = from_accession_to_id(segment)\n",
    "        version = version + from_id_to_version_accession(id)\n",
    "\n",
    "    for j in range(k_time):\n",
    "        id = k_data[i*20:(i+1)*20]\n",
    "        version = version + from_id_to_version_accession(id)\n",
    "\n",
    "    with open('including_version_{}.txt'.format(genus), 'w') as m:\n",
    "        m.writelines(list(map(lambda x:x+'\\n',version)))\n",
    "\n",
    "    os.system(\"grep -oP '[A-Z]+\\d+.\\d' including_version_{}.txt >> criterion_format_{}.txt\".format(genus,genus))\n",
    "    os.system(\"sort -u criterion_format_{}.txt > finally_{}.txt\".format(genus,genus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_id_to_version_accession(id):\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text\n",
    "\n",
    "    version_pattern = re.compile(r'>[A-Z]{1,2}\\d+.\\d')\n",
    "    version_info = re.findall(version_pattern, fasta_info)\n",
    "    return version_info\n",
    "\n",
    "with open('pure_num_alphacoronavirus.txt', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "data = list(map(lambda x:x.strip('\\n'),lines))\n",
    "time = len(data) // 20 + 1\n",
    "for i in range(time):\n",
    "    id = data[i*20:(i+1)*20]\n",
    "    print(from_id_to_version_accession(id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "\n",
    "def get_url_info_id(segment,num):\n",
    "    # Using requests function get the id from term\n",
    "    term_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&api_key=6e2f58038b074d964bb8b87f43623af7b508&term=' + \",\".join(segment)\n",
    "    term_info = requests.get(term_url, timeout=(5,5)).text\n",
    " \n",
    "    id_pattern = re.compile(r'<Id>(\\d+)</Id>')\n",
    "    accession_id = re.findall(id_pattern, term_info)\n",
    "\n",
    "    # get sequencing information from id\n",
    "    fasta_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=fasta&id={}&api_key=6e2f58038b074d964bb8b87f43623af7b508'.format(','.join(accession_id))\n",
    "    fasta_info = requests.get(fasta_url,timeout=(5,5)).text"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8 (default, Apr 13 2021, 19:58:26) \n[GCC 7.3.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "eace1a3d2d18dcb98353cfeb529f20792ec64dcc10ba0bdd358326e3f660811b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
