{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import gseapy as gp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "deg = pd.read_csv(\"/mnt/c/Users/ouyangkang/Downloads/all_diff_peak_se.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>idx</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>FDR</th>\n",
       "      <th>MeanDiff</th>\n",
       "      <th>peakName</th>\n",
       "      <th>se_id</th>\n",
       "      <th>nearestGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>12191</td>\n",
       "      <td>61168938</td>\n",
       "      <td>61169438</td>\n",
       "      <td>1.223151</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>0.593956</td>\n",
       "      <td>chr9:61168938-61169438</td>\n",
       "      <td>SE_12_025400461</td>\n",
       "      <td>I730028E13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr8</td>\n",
       "      <td>14433</td>\n",
       "      <td>89404193</td>\n",
       "      <td>89404693</td>\n",
       "      <td>1.644044</td>\n",
       "      <td>0.001825</td>\n",
       "      <td>0.551949</td>\n",
       "      <td>chr8:89404193-89404693</td>\n",
       "      <td>SE_12_025400088</td>\n",
       "      <td>Sall1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>14426</td>\n",
       "      <td>89392632</td>\n",
       "      <td>89393132</td>\n",
       "      <td>1.035705</td>\n",
       "      <td>0.010923</td>\n",
       "      <td>0.552092</td>\n",
       "      <td>chr8:89392632-89393132</td>\n",
       "      <td>SE_12_025400088</td>\n",
       "      <td>Sall1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>chr8</td>\n",
       "      <td>14431</td>\n",
       "      <td>89396602</td>\n",
       "      <td>89397102</td>\n",
       "      <td>1.944952</td>\n",
       "      <td>0.025567</td>\n",
       "      <td>0.374053</td>\n",
       "      <td>chr8:89396602-89397102</td>\n",
       "      <td>SE_12_025400088</td>\n",
       "      <td>Sall1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>chr13</td>\n",
       "      <td>6280</td>\n",
       "      <td>43778204</td>\n",
       "      <td>43778704</td>\n",
       "      <td>1.262150</td>\n",
       "      <td>0.084380</td>\n",
       "      <td>0.318489</td>\n",
       "      <td>chr13:43778204-43778704</td>\n",
       "      <td>SE_12_045400188</td>\n",
       "      <td>Cd83</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 seqnames    idx     start       end    Log2FC       FDR  \\\n",
       "0           1     chr9  12191  61168938  61169438  1.223151  0.000340   \n",
       "1           2     chr8  14433  89404193  89404693  1.644044  0.001825   \n",
       "2           3     chr8  14426  89392632  89393132  1.035705  0.010923   \n",
       "3           4     chr8  14431  89396602  89397102  1.944952  0.025567   \n",
       "4           5    chr13   6280  43778204  43778704  1.262150  0.084380   \n",
       "\n",
       "   MeanDiff                 peakName            se_id    nearestGene  \n",
       "0  0.593956   chr9:61168938-61169438  SE_12_025400461  I730028E13Rik  \n",
       "1  0.551949   chr8:89404193-89404693  SE_12_025400088          Sall1  \n",
       "2  0.552092   chr8:89392632-89393132  SE_12_025400088          Sall1  \n",
       "3  0.374053   chr8:89396602-89397102  SE_12_025400088          Sall1  \n",
       "4  0.318489  chr13:43778204-43778704  SE_12_045400188           Cd83  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>idx</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>FDR</th>\n",
       "      <th>MeanDiff</th>\n",
       "      <th>peakName</th>\n",
       "      <th>se_id</th>\n",
       "      <th>nearestGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2895</th>\n",
       "      <td>2896</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11632</td>\n",
       "      <td>84882943</td>\n",
       "      <td>84883443</td>\n",
       "      <td>-5.439819</td>\n",
       "      <td>0.059347</td>\n",
       "      <td>-0.255457</td>\n",
       "      <td>chr1:84882943-84883443</td>\n",
       "      <td>SE_12_045700798</td>\n",
       "      <td>Trip12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2894</th>\n",
       "      <td>2895</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2591</td>\n",
       "      <td>12032347</td>\n",
       "      <td>12032847</td>\n",
       "      <td>-5.317884</td>\n",
       "      <td>0.045216</td>\n",
       "      <td>-0.234265</td>\n",
       "      <td>chr16:12032347-12032847</td>\n",
       "      <td>SE_12_045700430</td>\n",
       "      <td>Shisa9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2920</th>\n",
       "      <td>2921</td>\n",
       "      <td>chr4</td>\n",
       "      <td>25638</td>\n",
       "      <td>143278255</td>\n",
       "      <td>143278755</td>\n",
       "      <td>-5.190113</td>\n",
       "      <td>0.342052</td>\n",
       "      <td>-0.213899</td>\n",
       "      <td>chr4:143278255-143278755</td>\n",
       "      <td>SE_12_045400191</td>\n",
       "      <td>Pdpn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2897</th>\n",
       "      <td>2898</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19794</td>\n",
       "      <td>117074485</td>\n",
       "      <td>117074985</td>\n",
       "      <td>-5.184113</td>\n",
       "      <td>0.115264</td>\n",
       "      <td>-0.212987</td>\n",
       "      <td>chr8:117074485-117074985</td>\n",
       "      <td>SE_12_045600282</td>\n",
       "      <td>Pkd1l2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2918</th>\n",
       "      <td>2919</td>\n",
       "      <td>chr11</td>\n",
       "      <td>26801</td>\n",
       "      <td>109671780</td>\n",
       "      <td>109672280</td>\n",
       "      <td>-5.096357</td>\n",
       "      <td>0.342052</td>\n",
       "      <td>-0.200062</td>\n",
       "      <td>chr11:109671780-109672280</td>\n",
       "      <td>SE_12_045500630</td>\n",
       "      <td>Prkar1a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 seqnames    idx      start        end    Log2FC       FDR  \\\n",
       "2895        2896     chr1  11632   84882943   84883443 -5.439819  0.059347   \n",
       "2894        2895    chr16   2591   12032347   12032847 -5.317884  0.045216   \n",
       "2920        2921     chr4  25638  143278255  143278755 -5.190113  0.342052   \n",
       "2897        2898     chr8  19794  117074485  117074985 -5.184113  0.115264   \n",
       "2918        2919    chr11  26801  109671780  109672280 -5.096357  0.342052   \n",
       "\n",
       "      MeanDiff                   peakName            se_id nearestGene  \n",
       "2895 -0.255457     chr1:84882943-84883443  SE_12_045700798      Trip12  \n",
       "2894 -0.234265    chr16:12032347-12032847  SE_12_045700430      Shisa9  \n",
       "2920 -0.213899   chr4:143278255-143278755  SE_12_045400191        Pdpn  \n",
       "2897 -0.212987   chr8:117074485-117074985  SE_12_045600282      Pkd1l2  \n",
       "2918 -0.200062  chr11:109671780-109672280  SE_12_045500630     Prkar1a  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgl_df = deg[deg['Log2FC'] >= 0].sort_values(by='Log2FC', ascending=False)\n",
    "mac_df = deg[deg['Log2FC'] <= 0].sort_values(by='Log2FC')\n",
    "mac_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_activated_mgl_genes = []\n",
    "top_mac_genes = []\n",
    "\n",
    "mgl_genes = mgl_df['nearestGene'].tolist()\n",
    "top_activated_mgl_genes += list(set(mgl_genes[:200]))\n",
    "\n",
    "if len(top_activated_mgl_genes) != 200:\n",
    "    for i in range(10000):\n",
    "        top_activated_mgl_genes += mgl_genes[i+201]\n",
    "        top_activated_mgl_genes = list(set(top_activated_mgl_genes))\n",
    "        if len(top_activated_mgl_genes) >= 200:\n",
    "            break\n",
    "\n",
    "mac_genes = mac_df['nearestGene'].tolist()\n",
    "top_mac_genes += list(set(mac_genes[:200]))\n",
    "\n",
    "if len(top_mac_genes) != 200:\n",
    "    for i in range(1000):\n",
    "        top_mac_genes += mac_genes[i+201]\n",
    "        top_mac_genes = list(set(top_mac_genes))\n",
    "        if len(top_mac_genes) >= 200:\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Pygb'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_mac_genes.pop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_10x_scRNA_2021',\n",
       " 'Allen_Brain_Atlas_down',\n",
       " 'Allen_Brain_Atlas_up',\n",
       " 'Azimuth_2023',\n",
       " 'Azimuth_Cell_Types_2021',\n",
       " 'BioCarta_2013',\n",
       " 'BioCarta_2015',\n",
       " 'BioCarta_2016',\n",
       " 'BioPlanet_2019',\n",
       " 'BioPlex_2017',\n",
       " 'CCLE_Proteomics_2020',\n",
       " 'CORUM',\n",
       " 'COVID-19_Related_Gene_Sets',\n",
       " 'COVID-19_Related_Gene_Sets_2021',\n",
       " 'Cancer_Cell_Line_Encyclopedia',\n",
       " 'CellMarker_2024',\n",
       " 'CellMarker_Augmented_2021',\n",
       " 'ChEA_2013',\n",
       " 'ChEA_2015',\n",
       " 'ChEA_2016',\n",
       " 'ChEA_2022',\n",
       " 'Chromosome_Location',\n",
       " 'Chromosome_Location_hg19',\n",
       " 'ClinVar_2019',\n",
       " 'DGIdb_Drug_Targets_2024',\n",
       " 'DSigDB',\n",
       " 'Data_Acquisition_Method_Most_Popular_Genes',\n",
       " 'DepMap_CRISPR_GeneDependency_CellLines_2023',\n",
       " 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
       " 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
       " 'Descartes_Cell_Types_and_Tissue_2021',\n",
       " 'Diabetes_Perturbations_GEO_2022',\n",
       " 'DisGeNET',\n",
       " 'Disease_Perturbations_from_GEO_down',\n",
       " 'Disease_Perturbations_from_GEO_up',\n",
       " 'Disease_Signatures_from_GEO_down_2014',\n",
       " 'Disease_Signatures_from_GEO_up_2014',\n",
       " 'DrugMatrix',\n",
       " 'Drug_Perturbations_from_GEO_2014',\n",
       " 'Drug_Perturbations_from_GEO_down',\n",
       " 'Drug_Perturbations_from_GEO_up',\n",
       " 'ENCODE_Histone_Modifications_2013',\n",
       " 'ENCODE_Histone_Modifications_2015',\n",
       " 'ENCODE_TF_ChIP-seq_2014',\n",
       " 'ENCODE_TF_ChIP-seq_2015',\n",
       " 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
       " 'ESCAPE',\n",
       " 'Elsevier_Pathway_Collection',\n",
       " 'Enrichr_Libraries_Most_Popular_Genes',\n",
       " 'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
       " 'Enrichr_Users_Contributed_Lists_2020',\n",
       " 'Epigenomics_Roadmap_HM_ChIP-seq',\n",
       " 'FANTOM6_lncRNA_KD_DEGs',\n",
       " 'GO_Biological_Process_2013',\n",
       " 'GO_Biological_Process_2015',\n",
       " 'GO_Biological_Process_2017',\n",
       " 'GO_Biological_Process_2017b',\n",
       " 'GO_Biological_Process_2018',\n",
       " 'GO_Biological_Process_2021',\n",
       " 'GO_Biological_Process_2023',\n",
       " 'GO_Cellular_Component_2013',\n",
       " 'GO_Cellular_Component_2015',\n",
       " 'GO_Cellular_Component_2017',\n",
       " 'GO_Cellular_Component_2017b',\n",
       " 'GO_Cellular_Component_2018',\n",
       " 'GO_Cellular_Component_2021',\n",
       " 'GO_Cellular_Component_2023',\n",
       " 'GO_Molecular_Function_2013',\n",
       " 'GO_Molecular_Function_2015',\n",
       " 'GO_Molecular_Function_2017',\n",
       " 'GO_Molecular_Function_2017b',\n",
       " 'GO_Molecular_Function_2018',\n",
       " 'GO_Molecular_Function_2021',\n",
       " 'GO_Molecular_Function_2023',\n",
       " 'GTEx_Aging_Signatures_2021',\n",
       " 'GTEx_Tissue_Expression_Down',\n",
       " 'GTEx_Tissue_Expression_Up',\n",
       " 'GTEx_Tissues_V8_2023',\n",
       " 'GWAS_Catalog_2019',\n",
       " 'GWAS_Catalog_2023',\n",
       " 'GeDiPNet_2023',\n",
       " 'GeneSigDB',\n",
       " 'Gene_Perturbations_from_GEO_down',\n",
       " 'Gene_Perturbations_from_GEO_up',\n",
       " 'Genes_Associated_with_NIH_Grants',\n",
       " 'Genome_Browser_PWMs',\n",
       " 'GlyGen_Glycosylated_Proteins_2022',\n",
       " 'HDSigDB_Human_2021',\n",
       " 'HDSigDB_Mouse_2021',\n",
       " 'HMDB_Metabolites',\n",
       " 'HMS_LINCS_KinomeScan',\n",
       " 'HomoloGene',\n",
       " 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
       " 'HuBMAP_ASCTplusB_augmented_2022',\n",
       " 'HumanCyc_2015',\n",
       " 'HumanCyc_2016',\n",
       " 'Human_Gene_Atlas',\n",
       " 'Human_Phenotype_Ontology',\n",
       " 'IDG_Drug_Targets_2022',\n",
       " 'InterPro_Domains_2019',\n",
       " 'Jensen_COMPARTMENTS',\n",
       " 'Jensen_DISEASES',\n",
       " 'Jensen_TISSUES',\n",
       " 'KEA_2013',\n",
       " 'KEA_2015',\n",
       " 'KEGG_2013',\n",
       " 'KEGG_2015',\n",
       " 'KEGG_2016',\n",
       " 'KEGG_2019_Human',\n",
       " 'KEGG_2019_Mouse',\n",
       " 'KEGG_2021_Human',\n",
       " 'KOMP2_Mouse_Phenotypes_2022',\n",
       " 'Kinase_Perturbations_from_GEO_down',\n",
       " 'Kinase_Perturbations_from_GEO_up',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_down',\n",
       " 'L1000_Kinase_and_GPCR_Perturbations_up',\n",
       " 'LINCS_L1000_CRISPR_KO_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_Consensus_Sigs',\n",
       " 'LINCS_L1000_Chem_Pert_down',\n",
       " 'LINCS_L1000_Chem_Pert_up',\n",
       " 'LINCS_L1000_Ligand_Perturbations_down',\n",
       " 'LINCS_L1000_Ligand_Perturbations_up',\n",
       " 'Ligand_Perturbations_from_GEO_down',\n",
       " 'Ligand_Perturbations_from_GEO_up',\n",
       " 'MAGMA_Drugs_and_Diseases',\n",
       " 'MAGNET_2023',\n",
       " 'MCF7_Perturbations_from_GEO_down',\n",
       " 'MCF7_Perturbations_from_GEO_up',\n",
       " 'MGI_Mammalian_Phenotype_2013',\n",
       " 'MGI_Mammalian_Phenotype_2017',\n",
       " 'MGI_Mammalian_Phenotype_Level_3',\n",
       " 'MGI_Mammalian_Phenotype_Level_4',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2019',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2021',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2024',\n",
       " 'MSigDB_Computational',\n",
       " 'MSigDB_Hallmark_2020',\n",
       " 'MSigDB_Oncogenic_Signatures',\n",
       " 'Metabolomics_Workbench_Metabolites_2022',\n",
       " 'Microbe_Perturbations_from_GEO_down',\n",
       " 'Microbe_Perturbations_from_GEO_up',\n",
       " 'MoTrPAC_2023',\n",
       " 'Mouse_Gene_Atlas',\n",
       " 'NCI-60_Cancer_Cell_Lines',\n",
       " 'NCI-Nature_2016',\n",
       " 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
       " 'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
       " 'NURSA_Human_Endogenous_Complexome',\n",
       " 'OMIM_Disease',\n",
       " 'OMIM_Expanded',\n",
       " 'Old_CMAP_down',\n",
       " 'Old_CMAP_up',\n",
       " 'Orphanet_Augmented_2021',\n",
       " 'PFOCR_Pathways',\n",
       " 'PFOCR_Pathways_2023',\n",
       " 'PPI_Hub_Proteins',\n",
       " 'PanglaoDB_Augmented_2021',\n",
       " 'Panther_2015',\n",
       " 'Panther_2016',\n",
       " 'PerturbAtlas',\n",
       " 'Pfam_Domains_2019',\n",
       " 'Pfam_InterPro_Domains',\n",
       " 'PheWeb_2019',\n",
       " 'PhenGenI_Association_2021',\n",
       " 'Phosphatase_Substrates_from_DEPOD',\n",
       " 'ProteomicsDB_2020',\n",
       " 'Proteomics_Drug_Atlas_2023',\n",
       " 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
       " 'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
       " 'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_AutoRIF_Gene_Lists',\n",
       " 'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_GeneRIF_Gene_Lists',\n",
       " 'Reactome_2013',\n",
       " 'Reactome_2015',\n",
       " 'Reactome_2016',\n",
       " 'Reactome_2022',\n",
       " 'Reactome_Pathways_2024',\n",
       " 'Rummagene_kinases',\n",
       " 'Rummagene_signatures',\n",
       " 'Rummagene_transcription_factors',\n",
       " 'SILAC_Phosphoproteomics',\n",
       " 'SubCell_BarCode',\n",
       " 'SynGO_2022',\n",
       " 'SynGO_2024',\n",
       " 'SysMyo_Muscle_Gene_Sets',\n",
       " 'TF-LOF_Expression_from_GEO',\n",
       " 'TF_Perturbations_Followed_by_Expression',\n",
       " 'TG_GATES_2020',\n",
       " 'TRANSFAC_and_JASPAR_PWMs',\n",
       " 'TRRUST_Transcription_Factors_2019',\n",
       " 'Table_Mining_of_CRISPR_Studies',\n",
       " 'Tabula_Muris',\n",
       " 'Tabula_Sapiens',\n",
       " 'TargetScan_microRNA',\n",
       " 'TargetScan_microRNA_2017',\n",
       " 'The_Kinase_Library_2023',\n",
       " 'The_Kinase_Library_2024',\n",
       " 'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
       " 'Tissue_Protein_Expression_from_ProteomicsDB',\n",
       " 'Transcription_Factor_PPIs',\n",
       " 'UK_Biobank_GWAS_v1',\n",
       " 'Virus-Host_PPI_P-HIPSTer_2020',\n",
       " 'VirusMINT',\n",
       " 'Virus_Perturbations_from_GEO_down',\n",
       " 'Virus_Perturbations_from_GEO_up',\n",
       " 'WikiPathway_2021_Human',\n",
       " 'WikiPathway_2023_Human',\n",
       " 'WikiPathways_2013',\n",
       " 'WikiPathways_2015',\n",
       " 'WikiPathways_2016',\n",
       " 'WikiPathways_2019_Human',\n",
       " 'WikiPathways_2019_Mouse',\n",
       " 'WikiPathways_2024_Human',\n",
       " 'WikiPathways_2024_Mouse',\n",
       " 'dbGaP',\n",
       " 'huMAP',\n",
       " 'lncHUB_lncRNA_Co-Expression',\n",
       " 'miRTarBase_2017']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp.get_library_name(organism='Mouse')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrichment(gene_list, file_name_prefix):\n",
    "    # enrichment analysis\n",
    "    enr = gp.enrich(gene_list,\n",
    "                    gene_sets=\"GO_Biological_Process_2023\",\n",
    "                    outdir=None)\n",
    "    # filtering p value is more than 0.05\n",
    "    result = enr.res2d[enr.res2d['P-value'] < 0.05]\n",
    "    # save result\n",
    "    result.to_csv(f\"/mnt/c/Users/ouyangkang/Desktop/{file_name_prefix}_result.csv\", index=None)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({\n",
    "    \"MAC\": top_mac_genes,\n",
    "    \"activated_MGL\": top_activated_mgl_genes\n",
    "})\n",
    "\n",
    "df.to_csv('/mnt/c/Users/ouyangkang/Desktop/diff_marker.csv', index=None, encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgl_result = enrichment(top_activated_mgl_genes[:100], 'activated_mgl_diff')\n",
    "mac_result = enrichment(top_mac_genes[:100], 'mac_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_mgl = mgl_result.iloc[[0, 1, 2 ,5, 8, 14, 15, 18, 19, 22]]\n",
    "select_mgl[\"Lable\"] = \"activated MGL\"\n",
    "select_mac = mac_result.iloc[[0, 1, 2, 3, 5, 6, 8, 16, 24, 39]]\n",
    "select_mac[\"Lable\"] = \"MAC\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
